{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genomics\n",
    "### Aims\n",
    " - gene catalogue\n",
    " - reference platform: functional and variation analyses\n",
    " - biodiversity and ecosystem (metagenomics)\n",
    "### Variation\n",
    " - larger genomes - more repetitive DNA\n",
    " - Eukaryotes - diploid (heterozygosity)\n",
    "### Gene Density\n",
    " - Human 12-15 genesmB\n",
    " - Animas mtDNA -14 genes in 14kb (1000 genes/mb)\n",
    "### NGS\n",
    "- sequencing by synthesis (Illumina)\n",
    "- single molecule real time (SMRT)\n",
    "- indexing - barcode\n",
    "- bridge amplification - cluster growth\n",
    "- illumina - Stargazing\n",
    "- PacBio or nanopore- Long reads in realtime - high error\n",
    "### Data structure\n",
    "- FASTA: starts with >, header\n",
    "fna, ffn, fasta/fas, faa, fq/fastq (quality files with PHRED quality scores)\n",
    "- Quality scores -log 10 (error probability)\n",
    "10 - 1 in 10 - 90% accuracy\n",
    "20 - 1 in 100 - 99% \n",
    "30 - 1 in 1000 - 99.9% accoracy\n",
    "- FASTQ files\n",
    "@title\n",
    "sequence line\n",
    "+optional repeat title\n",
    "quality line ASCII\n",
    "- Nucleic acid codes\n",
    "K G/T keto\n",
    "M A/C amino\n",
    "B G/T/C\n",
    "V G/C/A (any)\n",
    "S G/C (strong)\n",
    "W A/T (weak)\n",
    "D G/A/T\n",
    "Y TC pyrimidine\n",
    "R G/A purine\n",
    "H A/C/T\n",
    "\"-\" gap\n",
    "- GFF: gene-finding format\n",
    "describing genes and features\n",
    "1 sequence\n",
    "2 source\n",
    "3 feature\n",
    "4 start\n",
    "5 end\n",
    "6 score\n",
    "7 strand\n",
    "8 frame\n",
    "9 attributes\n",
    "- SAM: sequence alignment map\n",
    "- BAM: binary version\n",
    "11 mandatory fields in every row\n",
    "1 Read name\n",
    "2 bitwise flag\n",
    "3 reference sequence\n",
    "4 start position of read in reference\n",
    "5 map quality\n",
    "6 CIGAR string\n",
    "7 paired end read reference\n",
    "8 position of mate\n",
    "9 distance between reads\n",
    "10 read sequence\n",
    "11 read quality\n",
    "\n",
    "### Sequencing and Assembling\n",
    "- Reads\n",
    "- Contigs\n",
    "- Supercontigs or scaffolds\n",
    "\n",
    "NGS assemblers\n",
    "k-mers using Bruijn Graphs (DBG)\n",
    "* It has a problem with polimorfisms\n",
    "\n",
    "### assesing assembly\n",
    "- N50 measure of average contig size\n",
    "- Depth of coverage\n",
    "- Completeness of the gene catalogue\n",
    "\n",
    "#### challenges for genome assembly\n",
    "- heterozygosity\n",
    "- paralogy vs allelism (duplication)\n",
    "Ho do you tell a duplicated gene from alleles of a gene?\n",
    "- Sequence complexity: sequence repeats or trasposable \n",
    "\n",
    "#### extrinsic challenges\n",
    "- quality of DNA (errors)\n",
    "- length of DNA sequence reads (the longer the better)\n",
    "- coverage: depth of coverage\n",
    "- memory intensive\n",
    "- software\n",
    "\n",
    "### Genome annotation and function\n",
    "- function of genes can predict with prokka\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "- Secure Shell: (SSH) communication method between computers.\n",
    "- From windows is more difficult to connect to a server. (Putty is an option)\n",
    "- Transfering files: cyberduck, filezilla, winscp, scp\n",
    "\n",
    "\n",
    "## Specific editor\n",
    "- nano: https://wiki.gentoo.org/wiki/Nano/Basics_Guide\n",
    "\n",
    "A FASTQ file normally uses four lines per sequence.\n",
    "\n",
    "- Line 1 begins with a '@' character and is followed by a series of information that uniquely identifies each read.\n",
    "\n",
    "@HWI-ST1097:127:C0W5VACXX:5:1101:4820:2124 1:N:0:CTCAGA\n",
    "@SRR534005.299 HWI-ST945:93:c02g4acxx:3:1101:16012:2225/1\n",
    "\n",
    "- Line 2 is the raw sequence. The lenght of this sequence should correspond to the length of the read requested.\n",
    "\n",
    "- Line 3 is always '+' for output from the Illumina.\n",
    "\n",
    "- Line 4 is a string of Ascii-encoded base quality scores."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "efetch -db=nuccore -format=fasta -id=CM000046.1 > seqs/CM000046.1.fa\n",
    "\n",
    "# Running trimmomatic in a loop\n",
    "for R1 in *R1*\n",
    "do\n",
    "   R2=${R1//R1_001_small.fastq/R2_001_small.fastq}\n",
    "   R1paired=${R1//.fastq/_paired.fastq.gz}\n",
    "   R1unpaired=${R1//.fastq/_unpaired.fastq.gz}\t\n",
    "   R2paired=${R2//.fastq/_paired.fastq.gz}\n",
    "   R2unpaired=${R2//.fastq/_unpaired.fastq.gz}\t\n",
    "   echo \"java -jar $WORK/IntMolModule/STG/practiceData/Trimmomatic-0.36/trimmomatic-0.36.jar PE -threads 32 -phred33 $R1 $R2 $R1paired $R1unpaired $R2paired $R2unpaired\" >> trimmomatic.cmds\n",
    "done\n",
    "cat trimmomatic.cmds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "Cat  | xargs Rscript --vanilla pipeline.R \n",
    "\n",
    "#Read the sample from the command line\n",
    "sranumber<-commandArgs(trailingOnly=TRUE) \n",
    "print(sranumber)\n",
    "#Function Idk yet\n",
    "execute<-function(x, outputfile=NA, intern=FALSE, quitOnError=FALSE){\n",
    "  if(!is.na(outputfile) &&file.exists(outputfile)){\n",
    "    cat(\"Output for step exists, skipping this step\\n\");\n",
    "    invisible(\"\")\n",
    "  }\n",
    "  cat(\"----\", x, \"\\n\"); res<-system(x, intern=intern); cat(\">>>>\", res[1], \"\\n\")\n",
    "  if(res[1] >=1){ \n",
    "    cat(\"Error external process did not finish\\n\\n\");\n",
    "    if(quitOnError) q(\"no\")\n",
    "  }\n",
    "}\n",
    "\n",
    "input.dir<-\"/home/dev105/projects/rawdata\"\n",
    "input.base<-sranumber[1]\n",
    "output.dir<-paste0(\"/home/dev105/projects/outputdata/\", input.base,\".aln\")\n",
    "\n",
    "#Create an output folder\n",
    "if(!file.exists(input.dir)){ dir.create(input.dir, recursive=TRUE) }\n",
    "if(!file.exists(output.dir)){ dir.create(output.dir, recursive=TRUE) }\n",
    "\n",
    "#STEP 0 - SRA Download and Compress\n",
    "setwd(input.dir)\n",
    "execute(paste0(\" -p\", input.base)\n",
    "execute(paste0(\"fasterq-dump -p --split-files \", input.base), paste0(input.base, \"_1.fastq\"))\n",
    "execute(paste0(\"bgzip \", input.base, \"_1.fastq\"), paste0(input.base, \"_1.fastq.gz\"))\n",
    "execute(paste0(\"bgzip \", input.base, \"_2.fastq\"), paste0(input.base, \"_2.fastq.gz\"))\n",
    "\n",
    "#STEP 1 - READ Trimming\n",
    "trim.files<-c(\n",
    "                  paste0(input.dir, \"/\", input.base,\"_1.fastq.gz\"),\n",
    "                  paste0(input.dir, \"/\", input.base,\"_2.fastq.gz\"),\n",
    "                  paste0(output.dir, \"/\", input.base,\"_1.P.fastq.gz\"),\n",
    "                  paste0(output.dir, \"/\", input.base,\"_1.U.fastq.gz\"),\n",
    "                  paste0(output.dir, \"/\", input.base,\"_2.P.fastq.gz\"),\n",
    "                  paste0(output.dir, \"/\", input.base,\"_2.U.fastq.gz\")\n",
    "                )\n",
    "trim.path<-\"/home/dev105/devapps/Trimmomatic-0.39\"\n",
    "trim.exec<-paste0(\"java -jar \", trim.path, \"/trimmomatic-0.39.jar\")\n",
    "trim.opts<-paste0(\"ILLUMINACLIP:\",trim.path,\"/adapters/TruSeq3-PE-2.fa:2:30:10\")\n",
    "trim.opts<-paste0(trim.opts, \"LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\")\n",
    "trim.cmd<-paste0(trim.exec, \"PE \", paste0(trim.files, collapse=\"\"), \"\", trim.opts)\n",
    "\n",
    "execute(trim.cmd, trim.files[3])\n",
    "\n",
    "q(\"no\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
