{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 1.Setup\n",
    "Verify python3 or install it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/dev105/mambaforge/bin/python3\n"
     ]
    }
   ],
   "source": [
    "which python3 git jq\n",
    "\n",
    "#if not installed for Linux\n",
    "sudo apt-get install python3 \n",
    "sudo apt-get install pip3\n",
    "sudo apt-get install git\n",
    "sudo apt-get install jq\n",
    "\n",
    "#if not installed for Mac\n",
    "brew install python3\n",
    "brew install pip3\n",
    "brew install git\n",
    "brew install jq\n",
    "\n",
    "#cloud\n",
    "pip3 install dxpy\n",
    "dx -h\n",
    "git clone https://github.com/laderast/bash_bioinfo_scripts/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Sources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "#tutorial\n",
    "#https://github.com/raynamharris/Shell_Intro_for_Transcriptomics/blob/master/lessons/03_ReadWriteMoveCopy.mdwhoami, pwd, cd, ls, ls -tlh(sort by time, show details, human readable format)\n",
    "#https://wikis.utexas.edu/display/bioiteam/Scott's+list+of+linux+one-liners\n",
    "https://people.duke.edu/~ccc14/duke-hts-2018/cliburn/Bash_in_Jupyter.html \n",
    "https://laderast.github.io/bash_for_bioinformatics/\n",
    "https://bioinformatics.ccr.cancer.gov/docs/b4b/Module1_Unix_Biowulf/Lesson1/\n",
    "https://practicalcomputing.org/about.html\n",
    "https://regex101.com/\n",
    "\n",
    "Bash is a interpreter based on sh (shell)\n",
    "locate database of path names\n",
    "\n",
    "\n",
    "ls -tlh $PathDirectory\n",
    "ls folder\n",
    "ls -a #hidden files\n",
    "ls -A #exclude current and parent directory\n",
    "ls -l #characteristics\n",
    "ls -lt #last modified\n",
    "ls -R #recursive\n",
    "ls -d */ #showing only directories\n",
    "ls -l | grep -E '^d'\n",
    "ls -l | grep -Ev '^d' #showing only files\n",
    "ls *ipynb #globbing\n",
    "ls *Text*\n",
    "ls text? #only one character\n",
    "\n",
    "\n",
    "full and relative paths /, ~~, ., .., cd, tab for autocompletion\n",
    "\\ split over multiple lines\n",
    "with | a break is not necessary\n",
    "\n",
    "# files\n",
    "cat file1 file2 #explore at the same time\n",
    "head -n 4 file\n",
    "head -4 file\n",
    "tail -n 10 file\n",
    "tail +3 file\n",
    "less file #view and scroll, q for quitting\n",
    "cp file1 file2 $-R recursive\n",
    "mkdir\n",
    "mv #move or rename\n",
    "rm #remove -r recursive, rmdir for empty, -rf including subdirectories\n",
    "echo #print on the screen\n",
    "> #output\n",
    ">> #append output\n",
    "\n",
    "#here doc to pass multiline EOF or END, \n",
    "#unquoted to use shell, - leading tab ignored \"indentation\"\n",
    "uppercase $PWD or $(PWD)\n",
    "cat << EOF > text.txt\n",
    "crear dir\n",
    "un documento\n",
    "EOF\n",
    "\n",
    "#touch file.txt creates an empty file\n",
    "\n",
    "\n",
    "#nagivation\n",
    "up down #previous next command\n",
    "control+k delete from the cursor to the end\n",
    "control+e move to the end\n",
    "control+a move to the beginning\n",
    "control+d delete the character under the cursor\n",
    "control+c kill the current run\n",
    "\n",
    "wc filename #word count -l lines -w words -c characters\n",
    "* #wildcard\n",
    "? #singlecharacter\n",
    "\n",
    "gunzip SRR*_1R2.fastq.gz\n",
    "sort -n file #sort by number, -r reverse, uniq (shortlist), uniq -c (counts)\n",
    "\n",
    "\n",
    "grep #global regular expression print, lines\n",
    "grep -w #limit to find words\n",
    "grep -w \"this is a pattern\" file\n",
    "grep -n #numbers the lines\n",
    "grep -c #counts\n",
    "grep -i #case insensitive\n",
    "'^It' #starts the line with It\n",
    "'[0-9]+' #line with numbers\n",
    "\n",
    "grep -v #print lines that do not match\n",
    "grep -o '[A-Z]*' #get rid of numbers\n",
    "grep -E #no need to escape special characters, -Ev invert option\n",
    "grep -r \"hello\" ./*txt #recursive\n",
    "grep --color\n",
    "grep -l #files\n",
    "ls -d */ #directories\n",
    "\n",
    "ls -l | grep -E '^d' #folders that start with d\n",
    "\n",
    "## find\n",
    "find . -name iris*\n",
    "find . -name \"*unix*ipynb\"\n",
    "find . -iname #case insensitive\n",
    "find . -not -path \"*ipynb_checkpoints/*\" -iname \"*unix*ipynb\"\n",
    "find . -name \"*[csv|txt]\" -maxdepth 1 #one depth \n",
    "-ctime +1 #1day ago \n",
    "-mtime -1 #modified <1d \n",
    "-mmin -15\n",
    "\n",
    "#a set\n",
    "[abc]\n",
    "[a-z]\n",
    "! #negates\n",
    "[!a-d]*\n",
    "[^A-Z] match anything not in the set\n",
    "#pre-defined classes\n",
    "['[:alpha:]'] #letters\n",
    "['[:digit:]'] #numbers\n",
    "['[:punct:]'] #comma or .\n",
    "\n",
    "grep -E '()'\n",
    "grep -E '[]'\n",
    "grep -E ab+c # one or more of the preceding character abbbbbbcd\n",
    "grep -E ab*cd # zero or more of the preceding character\n",
    "grep -E 'ab{1,5}cd' #number of repetitions of the preceding character set\n",
    "'\\<the\\>' # word boundaries\n",
    "grep -E '([0-9]+).*\\1'\n",
    "'([0-9]+)_([0-9]+)_\\1_\\2' #repeated twice\n",
    "\n",
    "\n",
    ". one character\n",
    "+ one or more of the preceding character\n",
    "* zero or more characters\n",
    "^ start of the line\n",
    "$ end of the line\n",
    "[a|b|c] a or b or c\n",
    "(cat|dog) cat or dog\n",
    "\n",
    "\n",
    "grep \"[2|4]\" #lines 2 and 4\n",
    "egrep\n",
    "\n",
    "cat file.txt | awk -F '\\t' '$1==\"Mt\"' | head -3\n",
    "cat file.txt | awk -F '\\t' '$1==\"2\"' {print $3} | sort | uniq -c\n",
    "cat docs/echo.txt | awk 'BEGIN {FS = \",\"} ; $5 == \"M\"' > docs/male.txt\n",
    "\n",
    "grep -A 1 ##will print an aditional line after matching\n",
    "\n",
    "cut extracts columns separated by delimiters (-d)\n",
    "cut -d \" \" -f 2 file.txt #column 2 or 1,3,5 or 1-5\n",
    "\n",
    "\n",
    "sed for substitution\n",
    "sed -n \"1~2p\" file.txt #-n not print, extract only odd numbered lines\n",
    "sed -n '2000,2005p' #number of lines\n",
    "\n",
    "s/pattern/replacement/g/\n",
    "s->substitute\n",
    "g globally or first pattern found\n",
    "[0-9][0-9]* regular expression for matching a number \n",
    "\\. escape\n",
    "[[:space:]] matches a single space\n",
    "sed \"s/[0-9][0-9]*\\.[[:space:]]//g\" file.txt\n",
    "\n",
    "sed to remove dupplicate characters\n",
    "sed -E \"s/(.)\\1*/\\1/g\" junk.txt\n",
    "\n",
    "convert DNA to RNA\n",
    "cat file.txt | tr T U\n",
    "#tr -d 'AT' (delete)\n",
    "tr '\\t' ','#tsv to csv\n",
    "tr a-z A-Z\n",
    "tr ACTG TGAC\n",
    "\n",
    "convert reverse complement\n",
    "grep -o '[ACTG]*' junk.txt | tr ACGT TGCA | rev\n",
    "grep '[CG]' -o file | wc -l #count only CG\n",
    "\n",
    "#find unique features\n",
    "zcat $DATA_DIR/$GTF | grep -v '^#' | cut -f 3 | sort | uniq\n",
    "#counts\n",
    "zcat $DATA_DIR/$GTF | grep -v '^#' | cut -f 3 | sort | uniq -c\n",
    "#first 3 exons\n",
    "zcat $DATA_DIR/$GTF | grep -v '^#' | awk '$3 == \"exon\"' | head -3\n",
    "#attributes\n",
    "zcat $DATA_DIR/$GTF | grep -v '^#' | awk '$3 == \"gene\"' | cut -f 9 > genes.txt\n",
    "\n",
    "\n",
    "extract odd lines\n",
    "sed -n \"1~2p\" file.txt\n",
    "zcat $DATA_DIR/$FASTQ | sed -n '2~4p;' | head | tr -d '\\n' #extract every 4th line\n",
    "\n",
    "uniq #recognize two or more adjacent lines\n",
    "uniq -c #how many lines are repeated \n",
    "uniq -w N #first N characters of every line to decide, necessary to sort first\n",
    "\n",
    "history | grep \"pwd\" #search in your history\n",
    "\n",
    "\\ continue to next line or pipe |\n",
    "\n",
    "cd (cd - previous directory)\n",
    "mkdir -p seqs/data (create a parent directory and intermediate directories, rmdir -p also to remove them)\n",
    "efetch -db=nuccore -format=fasta -id=CM000046.1 > seqs/CM000046.1.fa\n",
    "\n",
    "\n",
    "gzip #to compress\n",
    "gunzip #to uncompress\n",
    "ls *fa | xargs -n1 bgzip \n",
    "zcat file.fa.gz | head\n",
    "tar #to compress directories (-cvzf or -xvzf to uncompress and recover)\n",
    "\tc compress\n",
    "\tx extract\n",
    "\tz gzip or gunzip\n",
    "\tf files\n",
    "\tv verbose\n",
    "\tt view list\n",
    "\n",
    "#integrity for data transfer checksum\n",
    "md5sum *fa > MD5_checksum\n",
    "#to check\n",
    "md5sum -c MD5_checksum\n",
    "#to compare then with the original file\n",
    "diff -u file1 file2 | cat\n",
    "\n",
    "set variables\n",
    "dirpath='PATH'\n",
    "fasta='file'\n",
    "echo ${dirpath} #use always ${}\n",
    "variables=$(pwd) #without spaces for commands\n",
    "echo \"hello ${NAME1}\"\n",
    "((NUM++)) double parantheses for mathematical expressions\n",
    "#seq function\n",
    "generates a range of numbers\n",
    "seq 3\n",
    "seq 2 5\n",
    "seq 5 2 9\n",
    "\n",
    "#if to check for file existence\n",
    "if [-f hello.txt]; then\n",
    "\tcat hello.txt\n",
    "else\n",
    "\techo \"no such file\"\n",
    "fi\n",
    "\n",
    "if [! -f \"data/file\"]; then\n",
    "\twget url -O data/file\n",
    "fi\n",
    "\n",
    "#test for conditional evaluation\n",
    "test -f file o [-f file]\n",
    "\n",
    "#Loops\n",
    "for i in SRRR*;\n",
    "do\n",
    "    echo ${i} #inserted in a line with other characters\n",
    "done\n",
    "for file in copy-SRR534005*; \n",
    "echo $file ${file//SRR534005/chicken}; \n",
    "do mv $file ${file//SRR534005/chicken}; \n",
    "done\n",
    "\n",
    "for FILE in $(ls *ipynb); do\n",
    "    echo $FILE\n",
    "done\n",
    "\n",
    "for ((i=0; i<5; i++)); do\n",
    "    echo $i\n",
    "done\n",
    "\n",
    "test -a = &&\n",
    "test -o = ||\n",
    "=~ to match regular expressions patterns\n",
    "\n",
    "for FILE in $(ls); do\n",
    "    if [[ \"${FILE}\" =~ ^Bash.*sh ]]; then\n",
    "        echo $FILE\n",
    "    fi\n",
    "done\n",
    "\n",
    "COUNTER=10\n",
    "while [ $COUNTER -gt 0 ]; do\n",
    "    echo $COUNTER\n",
    "    COUNTER=$(($COUNTER - 1))\n",
    "done\n",
    "\n",
    "-lt #less than\n",
    "-gt #greater than\n",
    "== #equaity\n",
    "!= #inequality\n",
    "\n",
    "#braces in loops\n",
    "for NUM in {000..005}; do\n",
    "    echo mkdir EXPT-${NUM}\n",
    "done\n",
    "echo foo.{c,cpp,h,hp}\n",
    "#loop first 10 records\n",
    "LINE_NUM=0\n",
    "while read LINE\n",
    "do\n",
    "    if [[ $LINE_NUM -lt 10 ]]; then\n",
    "        echo $LINE\n",
    "        (( LINE_NUM++ ))\n",
    "    fi\n",
    "done < genes.txt\n",
    "\n",
    "#command expressions $(command), #basename without path, dirname\n",
    "#remove the file extension Y=${x%.*}\n",
    "for FILE in $(ls ~/docs/*)\n",
    "do\n",
    "    DIR_NAME=$(dirname $FILE)\n",
    "    FILE_NAME=$(basename $FILE)\n",
    "    NAME=${FILE_NAME%.*}\n",
    "    NEW_DIR=$DIR_NAME/$NAME\n",
    "    NEW_FILE=${NAME}-copy.txt\n",
    "    mkdir -p $NEW_DIR\n",
    "    cat $FILE | head -3 | tail -2  > $NEW_DIR/$NEW_FILE\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "for FILE in $(ls $DATA_DIR/*_MA_J_S20_*gz)\n",
    "do\n",
    "    FIRST=$(zcat $FILE 2>/dev/null | head -1)\n",
    "    echo $(basename $FILE)\n",
    "    echo $FIRST\n",
    "    echo\n",
    "done\n",
    "#script\n",
    "#!/path/to/shell or bash path/file\n",
    "#!/bin/bash or which bash\n",
    "\n",
    "chmod +x /path/to/script\n",
    "\n",
    "#!/bin/bash\n",
    "# \"#\" gives you the number from the command line\n",
    "echo $# \n",
    "for ARG in {0..$#}; do\n",
    "    echo $ARG\n",
    "done\n",
    "EOF\n",
    "\n",
    "\n",
    "bash echo.sh\n",
    "for file in *.fastq\n",
    "do\n",
    "\techo ${file}\n",
    "done\n",
    "\n",
    "#script with input specifications\n",
    "bash wordcount.sh *R1.fastq\n",
    "for file in ${@}\n",
    "do\n",
    "\twc ${file}\n",
    "done\n",
    "#structure\n",
    "bin, data, results, manuscript\n",
    "\n",
    "\n",
    "cat > count_lines.sh << 'EOF'\n",
    "#!/bin/bash\n",
    "\n",
    "for FILE in $(ls)\n",
    "do\n",
    "    wc -l \"${FILE}\"\n",
    "done\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Loops\n",
    "\n",
    "  *  `for` is use for iterating over a fixed number of items (may be unknown at time of coding).\n",
    "  *  `while` is use for iterating until a certain condition is met\n",
    "\n",
    "#### Conditionals (for flow control)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#### Loops\n",
    "##### Examples of for\n",
    "for file in ../data/${@}\n",
    "do\n",
    "\twc ${file} >> ../data/wordcount.txt\n",
    "done\n",
    "\n",
    "for file in ../data/${@}\n",
    "do\n",
    "\toutput=$(basename ${file} .fastq)-wordcount.txt \n",
    "\twc ${file} >> ../results/${output}\n",
    "done\n",
    "\n",
    "for loops\n",
    "for i in $( ls ); do echo item: $i \n",
    "done\n",
    "\n",
    "##### Examples of while\n",
    "while true; do echo ‘hello’ done\n",
    "\n",
    "#### Conditionals\n",
    "if [ \"foo\" = \"foo\" ]; then\n",
    "else\n",
    "fi\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sources for scripts\n",
    "https://jeroenjanssens.com/dsatcl/\n",
    "https://missing.csail.mit.edu/2020/course-shell/\n",
    "https://swcarpentry.github.io/shell-novice/04-pipefilter/index.html https://datascienceatthecommandline.com/2e/chapter-2-getting-started.html?q=stdin#combining-command-line-tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "##### Iterating with xargs -I one by one -0 avoid blanks\n",
    "ls *.fasta | xargs -I% sh -c 'head %; echo \"\\n---\\n\"'\n",
    "#sh -c \"to call shell instead of bash\"\n",
    "#https://www.baeldung.com/linux/xargs-multiple-arguments\n",
    "my_variable=\"ggplot2\"\n",
    "\n",
    "echo \"My favorite R package is ${my_variable}\"\n",
    "my_var=\"chr1\"\n",
    "echo \"${my_var}_1.vcf.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "#script\n",
    "\n",
    "#STDIN -> STDOUT+STDERR\n",
    "I/O redirections: |,<,>\n",
    "ps -aef #show a list of all processes running\n",
    "bash sam_run my_file.bam out_stats.txt\n",
    "#!/bin/bash/\n",
    "samtools stats $1 > $2\n",
    "#!/bin/bash/\n",
    "samtools stats $2 > $1 #output primero"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#to hack SIGPIPE error in Jupyter\n",
    "cleanup () {\n",
    "    :\n",
    "}\n",
    "\n",
    "trap \"cleanup\" SIGPIPE\n",
    "#safety\n",
    "set -u #unset variables as an error and exit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "https://github.com/raynamharris/Shell_Intro_for_Transcriptomics/tree/master\n",
    "https://practicalcomputing.org/\n",
    "https://explainshell.com/\n",
    "A FASTQ file normally uses four lines per sequence.\n",
    "\n",
    "Line 1 begins with a '@' character and is followed by a series of information that uniquely identifies each read. Thus, this is our sequence. Here are two example:\n",
    "\n",
    "@HWI-ST1097:127:C0W5VACXX:5:1101:4820:2124 1:N:0:CTCAGA\n",
    "@SRR534005.299 HWI-ST945:93:c02g4acxx:3:1101:16012:2225/1\n",
    "\n",
    "Line 2 is the raw sequence. The lenght of this sequence should correspond to the length of the read requested (i.e. 75 bp, 150 bp). These are AGTC or N uppercase characters.\n",
    "\n",
    "Line 3 is always '+' for output from the Illumina machines in the Genomic Sequencing and Analysis Facility.\n",
    "\n",
    "Line 4 is a string of Ascii-encoded base quality scores encodes the quality values for the sequence in Line 2. This string contain the same number of symbols as letters in the sequence on Line 2.\n",
    "UNIX es un sistema operativo desarrollado en 1969. Se caracteriza principalmente por:\n",
    "Advertencia Utilizare terminos sin traducir por que considero que son importantes para seguir entendiendo adecuadamente Linux.\n",
    "UNIX filosofia\n",
    "    - El uso de texto plano para almacenamiento.\n",
    "    - Sistema jerarquico de archivos del sistema.\n",
    "    - Los dispositivos e IPC son tratados como archivos.\n",
    "    - Multiples programas pueden ser ejecutados juntos usando \"pipes\" en el interprete (a partir de ahora command line).\n",
    "Terminal Dispositivo usado para enviar y recibir datos del computador.\n",
    "Terminal.app Es un emulados del sistema operativo que permite interactuar con la computadora mediante el command line.\n",
    "Shell Es el programa que permite al sistema UNIX entender los comandos. *Usaremos bash, the bourne again shell*\n",
    "\n",
    "\n",
    "Commandos sencillos\n",
    "pwd\n",
    "ls\n",
    "cd\n",
    "cp\n",
    "mv\n",
    "mkdir\n",
    "rmman\n",
    "cat/zcat\n",
    "echo\n",
    "\n",
    "wildcards o caracteres propios de shell\n",
    "* matchs all\n",
    "? matchs a single character\n",
    "[A-Z]\n",
    "/\n",
    "./\n",
    "../\n",
    "~\n",
    "\n",
    "absolute path\n",
    "cd /usr/local/bin\n",
    "relative path\n",
    "cd ../lib\n",
    "\n",
    "\n",
    "Practica\n",
    "cd\n",
    "pwd\n",
    "ls -tl\n",
    "mkdir -p test-dir/lab1\n",
    "cd test-dir/lab1\n",
    "mkdir temp1 temp2 temp3\n",
    "mv temp1 temp2 temp3\n",
    "rm -rf temp*\n",
    "cd ..\n",
    "cd -\n",
    "pwd\n",
    "\n",
    "\n",
    "Brace expansion\n",
    "cho dog-{gone,bowl,bark}\n",
    "dog-gone dog-bowl dog-bark\n",
    "\n",
    " mkdir -p zmays-snps/{data/seqs,scripts,analysis}\n",
    "-p crear todos los subdirectorios necesarios\n",
    "Shell tiene ademas las funciones de autocompletar e historia\n",
    "la tecla tab (completara el comando o el nombre del archivo)\n",
    "Para acceder a la historia command se puede usar las teclas arriba o abajo.\n",
    "standard out to a file > o >> (>> agregara)\n",
    "\n",
    "por ejemplo concatenar fasta files\n",
    "cat *.fa > list.fa\n",
    "\n",
    "crear un log de errores\n",
    "command &> logfile\n",
    "command &>> logfile\n",
    "\n",
    "2> for errors\n",
    "\n",
    "tail-f (redirigir el output a la pantalla)\n",
    "\n",
    "input redirection <\n",
    "-single dash argument (standard input)\n",
    "\n",
    "Unix Pipes\n",
    "redirigir outputs entre programas\n",
    "El resultado de el programa X sera considerado input del programa Y\n",
    "Imagina descargar y analizar un archivo usando el metodo clasico de windows, tendria que crear muchos pasos e interactuar en todos ellos. Usando Unix puedo escribir todo y automatizarlo en un solo codigo.\n",
    "\n",
    "grep -v \"^>\" ../data/tb1.fasta | grep --color -i \"[^ATCG]\"\n",
    "CCCCAAAGACGGACCAATCCAGCAGCTTCTACTGCTAYCCATGCTCCCCTCCCTTCGCCGCCGCCGACGC\n",
    "\n",
    "\n",
    "he Unix program tee diverts a copy of your pipeline’s standard output stream to an intermediate file while still passing it through its standard output:\n",
    "program1 input.txt | tee intermediate-file.txt | program2 > results.txt\n",
    "\n",
    "We can tell the Unix shell to run a program in the background by appending an ampersand (&) to the end of our command. For example:\n",
    "\n",
    "program1 input.txt > results.txt &\n",
    "control + Z \n",
    "enviarlo a background bg\n",
    "shell muestra el numero ID o PID\n",
    "jobs\n",
    "traerlo de vuelta fg\n",
    "fg %<numero>\n",
    "\n",
    "aniquilar el proceso\n",
    "Control + C\n",
    "\n",
    "kill -9 <PID>\n",
    "kill -9 %<num> numero de trabajo\n",
    "\n",
    "\n",
    "Unix programs exit with an exit status, which indicates whether a program terminated without a problem or with an error. By Unix standards, an exit status of 0 indicates the process ran successfully, and any nonzero status indicates some sort of error has occurred (and hopefully the program prints an understandable error message, too). The exit status isn’t printed to the terminal, but your shell will set its value to a shell variable named $?. We can use the echo command to look at this variable’s value after running a command:\n",
    "program1 input.txt > results.txt; echo $?\n",
    "The shell provides two operators that implement this: one operator that runs the subsequent command only if the first command completed successfully (&&), and one operator that runs the next command only if the first completed unsuccessfully (||).\n",
    "\n",
    "For example, the sequence program1 input.txt > intermediate-results.txt && program2 intermediate-results.txt > results.txt will execute the second command only if previous commands have completed with a nonzero exit status.\n",
    "program1 input.txt > intermediate-results.txt || echo \"warning: an error occurred\" will print the message if error has occurred.\n",
    "True\n",
    "echo $?\n",
    "0\n",
    "false\n",
    "echo $?\n",
    "1\n",
    "true && echo \"works\"\n",
    "works\n",
    "true || echo \"works\"  (no saldra nada, solo si falla)\n",
    "false && echo \"first command was a success\"\n",
    "\n",
    "&& solo avisa cuando se cumple\n",
    "|| solo avisa cuando no se cumple\n",
    "\n",
    "command substitution\n",
    "echo \"There are $(grep -c '^>' ../data/improper.fa) entries in my FASTA file.\"\n",
    "There are 2 entries in my FASTA file.\n",
    "\n",
    "mkdir results-$(date +%F) #or\n",
    "mkdir results-`date +%F`\n",
    "\n",
    "RMarkdown\n",
    "https://data-skills.github.io/unix-and-bash/01-unix-git-intro/index.html\n",
    "\n",
    "# Shell\n",
    "command line que te permite controlar la computadora desde el teclado.\n",
    "reproducible, eficiente y rapido.\n",
    "Trabajar con cloud computers\n",
    "\n",
    "# whoami\n",
    "# pwd\n",
    "# ls\n",
    "ls -t #ordenado por tiempo\n",
    "ls -l #detalles\n",
    "ls -h #tamano\n",
    "\n",
    "# cd \n",
    "change directory\n",
    " relative path\n",
    " full path\n",
    " root directory\n",
    " home directory\n",
    " . este directorio\n",
    " .. un directorio arriba\n",
    " ../.. dos directorios arriba\n",
    " ~ home directorio\n",
    " cd #sin nada -> home directory\n",
    "\n",
    "# head y tail #\n",
    "muestra 10 primeras lineas\n",
    "head -n 4\n",
    "tail -n 4\n",
    "# less #\n",
    "muestra y navega en el contenido del archivo\n",
    "usar q para salir\n",
    "\n",
    "# cp #\n",
    "cp origen destino\n",
    "mv mover o renombrar\n",
    "\n",
    "# rm #\n",
    "eliminar\n",
    "-r directorios\n",
    "rmdir directorios\n",
    "\n",
    "# echo #\n",
    "mostrar en la pantalla\n",
    "echo 'hola!'\n",
    "echo -e \"3 Hello\\n4 again\" >> hello.txt\n",
    "\n",
    "2> #output and stderr\n",
    "\n",
    "# echo > y >> \n",
    ">> se agrega al archivo existente\n",
    "\n",
    "# cat\n",
    "muestra el contenido, pero tambien concatena varios archivos\n",
    "\n",
    "# Navegacion #\n",
    "\n",
    "control + d : eliminar el caracter en el cursor\n",
    "control + a : al inicio de la linea\n",
    "control + e : al final de la linea\n",
    "arriba : comando previo\n",
    "abajo : siguiente comando\n",
    "control + k : elimina desde el cursor al final de la linea\n",
    "control + c : forzar cerrar una tarea (salir)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# wc #\n",
    "Significa: \"word count\"\n",
    "** Uso: cuenta lineas, palabras o caracteres. **\n",
    "wc -l #numero de lineas\n",
    "wc -w #numero de palabras\n",
    "wc -c #numero de caracteres\n",
    "\n",
    "wc -l *.fastq > lenghts.txt\n",
    "\n",
    "sort -n lenghts.txt #ordena de forma numerica, -t',' -k 1 first column separated by commas\n",
    "head -n lenghts.txt\n",
    "wc *.fastq | sort -n | head -1\n",
    "\n",
    "# * #\n",
    "wildcard 0 o mas caracteres\n",
    "# ? #\n",
    "wildcard 1 caracter\n",
    "\n",
    "# gunzip #\n",
    "unzip\n",
    "\n",
    "https://wikis.utexas.edu/display/bioiteam/Scott's+list+of+linux+one-liners\n",
    "## grep ##\n",
    "Significa literalmente: “global/regular expression/print”\n",
    "\n",
    "**Uso: busca patrones (palabras o caracteres)y muestra las lineas que coinciden.**\n",
    "\n",
    "grep [argumento] [patron] [archivo]\n",
    "grep fastq lista.txt\n",
    "grep -w day haiku.txt (word or pattern \"\")\n",
    "grep -n it haiku.txt (numbers the lines)\n",
    "grep -i case insensitive\n",
    "grep -v do not match\n",
    "grep -A 1 also prints one line after\n",
    "\n",
    "uniq recognize two or more adjacent lines are the same\n",
    "uniq -c para contar\n",
    "uniq -w N solo usa los xxx caracteres iniciales para evaluar si son iguales\n",
    "sort haiku.txt | uniq -c \n",
    "\n",
    "history historial\n",
    "history | grep \"pwd\"\n",
    "\n",
    "# Bucles (Loops) #\n",
    "for -> repite el/los comandos para cada uno de los elementos de una lista\n",
    "\n",
    "for i in *.fastq\n",
    "do\n",
    "ls $i\n",
    "done\n",
    "\n",
    "$i -> variable\n",
    "do -> le dice a shell que daremos comandos\n",
    "ls -> comando o echo\n",
    "done -> fin\n",
    "\n",
    "$ for file in *.fastq\n",
    "> do\n",
    ">    echo \" fastq $file\"\n",
    "> done\n",
    "\n",
    "Nota: comilla y doble comilla -> doble comilla invocara un comando y reemplazara la variable con el archivo\n",
    "\n",
    "Nota: echo ${variable}\n",
    "echo 'esto!'\n",
    "echo esto\n",
    "\n",
    "Por ejemplo, si se necesita todos los R1\n",
    "$ ls *R1*\n",
    "\n",
    "$ for R1 in *R1*\n",
    "> do\n",
    ">    echo $R1\n",
    "> done\n",
    "\n",
    "Ahora obtener la R2\n",
    "newvariablename=${oldvariablename//oldtext/newtext}\n",
    "\n",
    "$ for R1 in *R1*\n",
    "> do\n",
    ">    R2=${R1//R1_001_small.fastq/R2_001_small.fastq}\n",
    ">    echo $R1 $R2\n",
    "> done\n",
    "\n",
    "$ for R1 in *R1*\n",
    "> do\n",
    ">    R2=${R1//R1_001_small.fastq/R2_001_small.fastq}\n",
    ">    echo \"do something to $R1 and also to $R2\n",
    "> done\n",
    "\n",
    "for R1 in *R1*\n",
    "do\n",
    "   R2=${R1//R1_001_small.fastq/R2_001_small.fastq}\n",
    "   R1paired=${R1//.fastq/_paired.fastq.gz}\n",
    "   R1unpaired=${R1//.fastq/_unpaired.fastq.gz}\t\n",
    "   R2paired=${R2//.fastq/_paired.fastq.gz}\n",
    "   R2unpaired=${R2//.fastq/_unpaired.fastq.gz}\t\n",
    "   echo \"java -jar $WORK/IntMolModule/STG/practiceData/Trimmomatic-0.36 PE -threads 32 -phred33 $R1 $R2 $R1paired $R1unpaired $R2paired $R2unpaired\"\n",
    "done\n",
    "\n",
    "\n",
    "#save it as cmds\n",
    "\n",
    "rm trimmomatic.cmds\n",
    "for R1 in *R1*\n",
    "do\n",
    "   R2=${R1//R1_001_small.fastq/R2_001_small.fastq}\n",
    "   R1paired=${R1//.fastq/_paired.fastq.gz}\n",
    "   R1unpaired=${R1//.fastq/_unpaired.fastq.gz}\t\n",
    "   R2paired=${R2//.fastq/_paired.fastq.gz}\n",
    "   R2unpaired=${R2//.fastq/_unpaired.fastq.gz}\t\n",
    "   echo \"java -jar $WORK/IntMolModule/STG/practiceData/Trimmomatic-0.36/trimmomatic-0.36.jar PE -threads 32 -phred33 $R1 $R2 $R1paired $R1unpaired $R2paired $R2unpaired\" >> trimmomatic.cmds\n",
    "done\n",
    "cat trimmomatic.cmds\n",
    "## Bash Scripts \n",
    "extension .sh\n",
    "son archivos ejecutables\n",
    "se ejecutan como bash archivo.sh\n",
    "\n",
    "\n",
    "for file in ${@}\n",
    "do \n",
    "    wc ${file}\n",
    "done\n",
    "\n",
    "Organizacion de proyectos\n",
    "bin\n",
    "data\n",
    "results\n",
    "manuscripts\n",
    "\n",
    "\n",
    "for file in ../data/${@}\n",
    "do\n",
    "    output=$(basename ${file} .fastq)-wordcount.txt\n",
    "    wc ${file} >> ../data/wordcount.txt\n",
    "done\n",
    "\n",
    "mkdir dataseq\n",
    "\n",
    "while read\n",
    "fasterq-dump -p --split-files $inputsra -O ./dataseq\n",
    "\n",
    "\n",
    "\n",
    "echo \"Start!\"\n",
    "\n",
    "for sra in (txt)\n",
    "do\n",
    "    echo \"${p}\"\n",
    "    while [ $(jobs | wc -l) -ge \"$maxproc\" ]\n",
    "    do\n",
    "        sleep 1\n",
    "    done\n",
    "    base=${fname%_R1*}\n",
    "    echo starting new job with ongoing=$(jobs | wc -l)\n",
    "    bwa-0.7.5a/bwa mem -t 4 human_g1k_v37.fasta \"${base}_R1.fastq.gz\" \"${base}_R2.fastq.gz\" >\"$destdir/${base}_R1_R2.sam\" &\n",
    "done\n",
    "\n",
    "\n",
    "#gunzip\n",
    "\n",
    "Compressing multiple files with gzip\n",
    "To compress multiple files with gzip, you can use *.txt wildcard to select all files with .txt extension in current directory.\n",
    "gzip *.txt\n",
    "This command will compress all files with .txt extension in current directory, creating new compressed files with .gz extension.\n",
    "Compressing files while keeping original file\n",
    "To compress a file with gzip while keeping original file, you can use -k option.\n",
    "gzip -k file.txt\n",
    "\n",
    "From <https://www.tutorialspoint.com/using-gzip-and-gunzip-in-linux> \n",
    "\n",
    "wget http://kandurilab.org/users/santhilal/courses/UNIX/materials/course_file_repo.zip\n",
    "\n",
    "Workspace: shell or shell prompt\n",
    "Linux - Ubuntu terminal - bash\n",
    "\n",
    "\n",
    "\n",
    "#list\n",
    "ls \n",
    "#detailed list (long list)\n",
    "ls -l \n",
    "ls -lh\n",
    "#pwd\n",
    "#mkdir\n",
    "#cp\n",
    "#mv (risky)\n",
    "#nano will create files\n",
    "Alt+ change between files\n",
    "Control o to save\n",
    "Control X exit\n",
    "sudo apt-get update\n",
    "\n",
    "\n",
    " 'file1.txt' -> 'myfile1.txt', 'file2.txt' -> 'myfile2.txt'\n",
    "Now let's merge the created text files using command 'cat' (concatenate).\n",
    "\n",
    "$$ cat myfile1 myfile2 > mergedfile.txt # This merges the contents of the file in a new file\n",
    "$$ cat mergedfile.txt # Display the content of merged file\n",
    "\n",
    "Rm myfile*\n",
    "Rm -mydummy\n",
    "Sudo apt-get install unzip\n",
    "\n",
    "Unzip file\n",
    "Head\n",
    "Less\n",
    "More\n",
    "\n",
    "Salir con q\n",
    "\n",
    "\n",
    "Count words or lines\n",
    "Wc -l *.list (list es el tipo de nombre)\n",
    "\n",
    "\n",
    "Match genes start with 'USP'.\n",
    "$$ cat H0.list | grep \"^USP*\" # Here '^' means 'line starts with'\n",
    "$$ cat H0.list | grep \"USP*\"\n",
    "Match genes ending with '.' followed by some digit/number.\n",
    "$$ cat H0.list | grep \"\\.[0-9]$\"\n",
    "\\. -> to scape of matching\n",
    "* suffix\n",
    ". Any digit\n",
    "$ the ending part\n",
    "Beginning of line ( ^ )\n",
    "Count of empty lines \n",
    "\"^$\"  FILE\n",
    "Grep -w \"….\" (only 4 characters)\n",
    "*. \n",
    "\\+ one or more ocurrences (sin considerar espacios) dentro de la oracion mas de un espacio\n",
    "Cero o no espacios *\n",
    "\\? Cero o una ocurrencia\n",
    "\\ escape the special character\n",
    "Character class [0-9]\n",
    "\n",
    "Exception in the character class grep -I \"^[^aeiou]\"\n",
    "\n",
    "Grep -B (lineas antes y despues)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "$$ cat hg38_gene_annotation_mart_export.txt | grep -v \"^#\" | cut -f1 | sort | uniq | wc -l # This counts total number of unique genes in the annotation\n",
    "$$ cat hg38_gene_annotation_mart_export.txt | grep -v \"^#\" | cut -f7 | sort | uniq | wc -l # This counts total number of unique transcripts\n",
    "$$ cat hg38_gene_annotation_mart_export.txt | grep -v \"^#\" | grep -w \"protein_coding\" | cut -f1 | sort | uniq | wc -l # Counts only total number of protein coding genes\n",
    "\n",
    "Zcat to open .gz files\n",
    "\n",
    "zcat Homo_sapiens.GRCh38.87_badStructure.table.gz | sed 's/gene_id \"//' | sed 's/gene_name \"//' | sed 's/gene_biotype \"//' | sed 's/\"//g' | sed '1i\\Geneid\\tGeneSymbol\\tChromosome\\tClass\\tStrand\\tLength' | sed 's/ //g' | head\n",
    "\n",
    "SED\n",
    "https://www.gnu.org/software/sed/manual/sed.html\n",
    "\n",
    "Sed is more efficient and make only one pass over the input (filter text in a pipeline)\n",
    "Sed $script INPUTFILE\n",
    "sed 's/hello/world/' input.txt > output.txt\n",
    "\n",
    "Alternatives:\n",
    "sed 's/hello/world/' input.txt > output.txt\n",
    "sed 's/hello/world/' < input.txt > output.txt\n",
    "cat input.txt | sed 's/hello/world/' - > output.txt\n",
    "\n",
    "Sobre el mismo archivo \n",
    "-i\n",
    "sed -i 's/hello/world/' input.txt #sobre el mismo archivo\n",
    "#otros: w s///w\n",
    "\n",
    "Use -n to suppress output, and the p command to print specific lines. The following command prints only line 45 of the input file:\n",
    "\n",
    "sed -n '45p' file.txt\n",
    "sed treats multiple input files as one long stream. The following example prints the first line of the first file (one.txt) and the last line of the last file (three.txt). Use -s to reverse this behavior.\n",
    "\n",
    "sed -n  '1p ; $p' one.txt two.txt three.txt\n",
    "\n",
    "Without -e or -f options, sed uses the first non-option parameter as the script, and the following non-option parameters as input files. If -e or -f options are used to specify a script, all non-option parameters are taken as input files. Options -e and -f can be combined, and can appear multiple times (in which case the final effective script will be concatenation of all the individual scripts).\n",
    "\n",
    "The following examples are equivalent:\n",
    "\n",
    "sed 's/hello/world/' input.txt > output.txt\n",
    "\n",
    "sed -e 's/hello/world/' input.txt > output.txt\n",
    "sed --expression='s/hello/world/' input.txt > output.txt\n",
    "\n",
    "echo 's/hello/world/' > myscript.sed\n",
    "sed -f myscript.sed input.txt > output.txt\n",
    "sed --file=myscript.sed input.txt > output.txt\n",
    "The full format for invoking sed is:\n",
    "\n",
    "sed OPTIONS... [SCRIPT] [INPUTFILE...]\n",
    "sed may be invoked with the following command-line options:\n",
    "\n",
    "--version\n",
    "Print out the version of sed that is being run and a copyright notice, then exit.\n",
    "\n",
    "--help\n",
    "Print a usage message briefly summarizing these command-line options and the bug-reporting address, then exit.\n",
    "\n",
    "-n\n",
    "--quiet\n",
    "--silent\n",
    "By default, sed prints out the pattern space at the end of each cycle through the script (see How sed works). These options disable this automatic printing, and sed only produces output when explicitly told to via the p command.\n",
    "\n",
    "--debug\n",
    "Print the input sed program in canonical form, and annotate program execution.\n",
    "\n",
    "$ echo 1 | sed '\\%1%s21232'\n",
    "3\n",
    "\n",
    "$ echo 1 | sed --debug '\\%1%s21232'\n",
    "SED PROGRAM:\n",
    "  /1/ s/1/3/\n",
    "INPUT:   'STDIN' line 1\n",
    "PATTERN: 1\n",
    "COMMAND: /1/ s/1/3/\n",
    "PATTERN: 3\n",
    "END-OF-CYCLE:\n",
    "3\n",
    "-e script\n",
    "--expression=script\n",
    "Add the commands in script to the set of commands to be run while processing the input.\n",
    "\n",
    "-f script-file\n",
    "--file=script-file\n",
    "Add the commands contained in the file script-file to the set of commands to be run while processing the input.\n",
    "\n",
    "-i[SUFFIX]\n",
    "--in-place[=SUFFIX]\n",
    "This option specifies that files are to be edited in-place. GNU sed does this by creating a temporary file and sending output to this file rather than to the standard output.1.\n",
    "\n",
    "This option implies -s.\n",
    "\n",
    "When the end of the file is reached, the temporary file is renamed to the output file’s original name. The extension, if supplied, is used to modify the name of the old file before renaming the temporary file, thereby making a backup copy2).\n",
    "\n",
    "This rule is followed: if the extension doesn’t contain a *, then it is appended to the end of the current filename as a suffix; if the extension does contain one or more * characters, then each asterisk is replaced with the current filename. This allows you to add a prefix to the backup file, instead of (or in addition to) a suffix, or even to place backup copies of the original files into another directory (provided the directory already exists).\n",
    "\n",
    "If no extension is supplied, the original file is overwritten without making a backup.\n",
    "\n",
    "Because -i takes an optional argument, it should not be followed by other short options:\n",
    "\n",
    "sed -Ei '...' FILE\n",
    "Same as -E -i with no backup suffix - FILE will be edited in-place without creating a backup.\n",
    "\n",
    "sed -iE '...' FILE\n",
    "This is equivalent to --in-place=E, creating FILEE as backup of FILE\n",
    "\n",
    "Be cautious of using -n with -i: the former disables automatic printing of lines and the latter changes the file in-place without a backup. Used carelessly (and without an explicit p command), the output file will be empty:\n",
    "\n",
    "# WRONG USAGE: 'FILE' will be truncated.\n",
    "sed -ni 's/foo/bar/' FILE\n",
    "-l N\n",
    "--line-length=N\n",
    "Specify the default line-wrap length for the l command. A length of 0 (zero) means to never wrap long lines. If not specified, it is taken to be 70.\n",
    "\n",
    "--posix\n",
    "GNU sed includes several extensions to POSIX sed. In order to simplify writing portable scripts, this option disables all the extensions that this manual documents, including additional commands. Most of the extensions accept sed programs that are outside the syntax mandated by POSIX, but some of them (such as the behavior of the N command described in Reporting Bugs) actually violate the standard. If you want to disable only the latter kind of extension, you can set the POSIXLY_CORRECT variable to a non-empty value.\n",
    "\n",
    "-b\n",
    "--binary\n",
    "This option is available on every platform, but is only effective where the operating system makes a distinction between text files and binary files. When such a distinction is made—as is the case for MS-DOS, Windows, Cygwin—text files are composed of lines separated by a carriage return and a line feed character, and sed does not see the ending CR. When this option is specified, sed will open input files in binary mode, thus not requesting this special processing and considering lines to end at a line feed.\n",
    "\n",
    "--follow-symlinks\n",
    "This option is available only on platforms that support symbolic links and has an effect only if option -i is specified. In this case, if the file that is specified on the command line is a symbolic link, sed will follow the link and edit the ultimate destination of the link. The default behavior is to break the symbolic link, so that the link destination will not be modified.\n",
    "\n",
    "-E\n",
    "-r\n",
    "--regexp-extended\n",
    "Use extended regular expressions rather than basic regular expressions. Extended regexps are those that egrep accepts; they can be clearer because they usually have fewer backslashes. Historically this was a GNU extension, but the -E extension has since been added to the POSIX standard (http://austingroupbugs.net/view.php?id=528), so use -E for portability. GNU sed has accepted -E as an undocumented option for years, and *BSD seds have accepted -E for years as well, but scripts that use -E might not port to other older systems. See Extended regular expressions.\n",
    "\n",
    "-s\n",
    "--separate\n",
    "By default, sed will consider the files specified on the command line as a single continuous long stream. This GNU sed extension allows the user to consider them as separate files: range addresses (such as ‘/abc/,/def/’) are not allowed to span several files, line numbers are relative to the start of each file, $ refers to the last line of each file, and files invoked from the R commands are rewound at the start of each file.\n",
    "\n",
    "--sandbox\n",
    "In sandbox mode, e/w/r commands are rejected - programs containing them will be aborted without being run. Sandbox mode ensures sed operates only on the input files designated on the command line, and cannot run external programs.\n",
    "\n",
    "-u\n",
    "--unbuffered\n",
    "Buffer both input and output as minimally as practical. (This is particularly useful if the input is coming from the likes of ‘tail -f’, and you wish to see the transformed output as soon as possible.)\n",
    "\n",
    "-z\n",
    "--null-data\n",
    "--zero-terminated\n",
    "Treat the input as a set of lines, each terminated by a zero byte (the ASCII ‘NUL’ character) instead of a newline. This option can be used with commands like ‘sort -z’ and ‘find -print0’ to process arbitrary file names.\n",
    "\n",
    "If no -e, -f, --expression, or --file options are given on the command-line, then the first non-option argument on the command line is taken to be the script to be executed.\n",
    "\n",
    "If any command-line parameters remain after processing the above, these parameters are interpreted as the names of input files to be processed. A file name of ‘-’ refers to the standard input stream. The standard input will be processed if no file names are specified.\n",
    "\n",
    "For printing specific range of lines from the file.\n",
    "\n",
    "$$ sed -n 1,2p Homo_sapiens.GRCh38.87_annotation.table\n",
    "$$ sed -n 4,6p Homo_sapiens.GRCh38.87_annotation.table\n",
    "For deleting lines\n",
    "\n",
    "$$ cat Homo_sapiens.GRCh38.87_annotation.table | sed '1d' | head\n",
    "\n",
    "Translate or replace\n",
    "Tr https://www.thegeekstuff.com/2012/12/linux-tr-command/\n",
    "$$ echo \"santhilal\" | tr a-z A-Z\n",
    "\n",
    "\n",
    "\n",
    "https://www.gnu.org/software/gawk/manual/html_node/Comparison-Operators.html\n",
    "\n",
    "https://genome.ucsc.edu/FAQ/FAQformat#format1\n",
    "\n",
    "\n",
    "pipe: combining different commands to get required output\n",
    "\n",
    "sudo apt-get upgrade -y\n",
    "\n",
    "sudo apt-get dist-upgrade -y\n",
    "\n",
    "sudo apt-get autoremove -y\n",
    "\n",
    "sudo apt-get autoclean -y\n",
    "\n",
    "\n",
    "Onedrive\n",
    "wget -qO - https://download.opensuse.org/repositories/home:/npreining:/debian-ubuntu-onedrive/xUbuntu_22.04/Release.key | gpg --dearmor | sudo tee /usr/share/keyrings/obs-onedrive.gpg > /dev/null\n",
    "\n",
    "echo \"deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/obs-onedrive.gpg] https://download.opensuse.org/repositories/home:/npreining:/debian-ubuntu-onedrive/xUbuntu_22.04/./\" | sudo tee /etc/apt/sources.list.d/onedrive.list\n",
    "\n",
    "sudo apt-get update\n",
    "\n",
    "sudo apt install --no-install-recommends --no-install-suggests onedrive\n",
    "\n",
    "\n",
    "#to get a detailed list of folders\n",
    "Ls -shl *\n",
    "\n",
    "\n",
    "Linux\n",
    "Turn windows features on or off\n",
    "check windows subsystem for linux\n",
    "follow the instructions and reboot\n",
    "microsoft store → ubuntu\n",
    "wsl --install\n",
    "Run a script\n",
    "\t1. Write the script file using\n",
    "nano script-name-here.sh\n",
    "\t2. Set execute permission on your script using chmod command :\n",
    "chmod +x script-name-here.sh\n",
    "\t3. To run your script :./script-name-here.shAnother option is as follows to execute shell script:sh script-name-here.shORbash script-name-here.sh\n",
    "“terminal”, “command”, “prompt” or “shell” or Ctrl-Alt-T.\n",
    "prompt, and you might sometimes see instructions that say “bring up a prompt”, “open a command prompt”, “at the bash prompt”\n",
    "pwd is an abbreviation of ‘print working directory’ (show the path)\n",
    "change directories (cd), another to list their contents (ls), a third to rename or move files (mv)\n",
    "cd /\n",
    "pwd\n",
    "cd home\n",
    "pwd\n",
    "cd ..\n",
    "pwd\n",
    "cd\n",
    "pwd\n",
    "cd ../.. #if you have to move up through multiple levels\n",
    "pwd\n",
    "cd ../../etc\n",
    "pwd\n",
    "cd /etc #absolute pathway\n",
    "pwd\n",
    "whoami\n",
    "cd /home/USERNAME/Desktop\n",
    "pwd\n",
    "#“starting from the root directory”\n",
    "cd ~\n",
    "pwd\n",
    "\n",
    "cd ~/Desktop\n",
    "pwd\n",
    "https://ubuntu.com/tutorials/command-line-for-beginners#4-creating-folders-and-files\n",
    "Mkdir - create files\n",
    "mkdir /tmp/tutorial\n",
    "cd /tmp/tutorial\n",
    "mkdir dir1 dir2 dir3\n",
    "mkdir -p dir4/dir5/dir6\n",
    "ls\n",
    "-p or --parent\n",
    "mkdir \"folder 1\"\n",
    "mkdir 'folder 2'\n",
    "mkdir folder\\\\ 3\n",
    "mkdir \"folder 4\" \"folder 5\"\n",
    "mkdir -p \"folder 6\"/\"folder 7\"\n",
    "ls\n",
    "ls > output.txt\n",
    "#we can use the cat command to look at its content:\n",
    "cat output.txt\n",
    "echo \"This is a test\"\n",
    "echo \"This is a test\" > test_1.txt\n",
    "echo \"This is a second test\" > test_2.txt\n",
    "echo \"This is a third test\" > test_3.txt\n",
    "ls\n",
    "cat test_1.txt test_2.txt test_3.txt\n",
    "question mark (”?”) can be used to indicate “any single character” within the file name. An asterisk (”*”) can be used to indicate “zero or more characters”. These are sometimes referred to as “wildcard” characters. A couple of examples might help, the following commands all do the same thing:\n",
    "\n",
    "cat test_1.txt test_2.txt test_3.txt\n",
    "cat test_?.txt\n",
    "cat test_*\n",
    "cat t*, meaning “concatenate all the files whose names start with a t and are followed by zero or more other characters”. Let’s use this capability to join all our files together into a single new file, then view it:\n",
    "\n",
    "cat t* > combined.txt\n",
    "cat combined.txt\n",
    "Up Arrow and Down Arrow keys to move back and forth through the history of commands you’ve used. Press the Up Arrow a couple of times to get to the first cat and press Enter to run it, then do the same again to get to the second.\n",
    "\n",
    "If you do want to append to, rather than replace, the content of the files, double up on the greater-than character:\n",
    "cat t* >> combined.txt\n",
    "echo \"I've appended a line!\" >> combined.txt\n",
    "cat combined.txt\n",
    "#to see more text\n",
    "less combined.txt \n",
    "Clear - clear the windows\n",
    "Mv “filepath” newname\n",
    "mv combined.txt dir1\n",
    "mv dir1/* .\n",
    "mv combined.txt ..\n",
    "mv combined.txt test_* dir3 dir2 ls ls dir2\n",
    "mv dir2/combined.txt dir4/dir5/dir6\n",
    "ls dir2\n",
    "ls dir4/dir5/dir6\n",
    "cp dir4/dir5/dir6/combined.txt .\n",
    "ls dir4/dir5/dir6\n",
    "ls\n",
    "cp combined.txt backup_combined.txt\n",
    "ls\n",
    "mv backup_combined.txt combined_backup.txt\n",
    "ls\n",
    "mv \"folder 1\" folder_1\n",
    "mv \"folder 2\" folder_2\n",
    "mv \"folder 3\" folder_3\n",
    "mv \"folder 4\" folder_4\n",
    "mv \"folder 5\" folder_5\n",
    "mv \"folder 6\" folder_6\n",
    "ls\n",
    "Chmod +x filename - make executable\n",
    "head .bashrc - go to the bash\n",
    "Cp .bashrc bashrc.bak\n",
    "Nano .bashrc\n",
    "Write at the end:\n",
    "export PATH=$PATH:”filepath”\n",
    "Rm - remove\n",
    "rm dir4/dir5/dir6/combined.txt combined_backup.txt\n",
    "Perhaps we should remove some of those excess directories as well:\n",
    "rmdir folder_*\n",
    "rmdir -p dir1/dir2/dir3\n",
    "rm -r folder_6\n",
    "Rm -rf director (remove not empty directory)\n",
    "ls\n",
    "If you’re at all uncertain use the -i (interactive) option to rm, which will prompt you to confirm the deletion of each file; enter Y to delete it, N to keep it, and press Ctrl-C to stop the operation entirely.\n",
    "\n",
    "How many lines are there in your combined.txt file? The wc (word count) command can tell us that, using the -l switch to tell it we only want the line count (it can also do character counts and, as the name suggests, word counts):\n",
    "wc -l combined.txt\n",
    "how many files and folders are in your home directory, and then tidy up after yourself, you could do this:\n",
    "ls ~ > file_list.txt\n",
    "wc -l file_list.txt\n",
    "rm file_list.txt\n",
    "piping the data from one command to another\n",
    "ls ~ | wc -l\n",
    "t’s better to use less to view it, and that advice still applies when using a pipe (remember, press q to quit):\n",
    "\n",
    "ls /etc | less\n",
    "tarUnix has a command, uniq, that will only output unique lines in the file.\n",
    "cat combined.txt | uniq | wc -l\n",
    "cat combined.txt | uniq | less\n",
    "\n",
    "su command. This is variously described as being short for ‘superuser’ or ‘switch user’\n",
    "logout command (or Ctrl-D shortcut)\n",
    "cat /etc/shadow\n",
    "sudo cat /etc/shadow\n",
    "apt or apt-get commands. If the instructions require you to first add a new software repository to your system, using the apt-add-repository command, by editing files in /etc/apt\n",
    "Shortcut to open the terminal in Ubuntu:\n",
    "\tControl+Alt+T\n",
    "Update Ubuntu and then install a software\n",
    "sudo apt update\n",
    "sudo apt install evolution\n",
    "sudo apt install evolution-ews\n",
    "tree\n",
    "cd /tmp/tutorial\n",
    "tree\n",
    "sudo su, be aware that every command after that will be running as the root user.\n",
    "dot (”.”) is enough to make it disappear.\n",
    "\n",
    "cd /tmp/tutorial\n",
    "ls\n",
    "mv combined.txt .combined.txt\n",
    "ls\n",
    "to continue working with them\n",
    "cat .combined.txt\n",
    "mkdir .hidden\n",
    "mv .combined.txt .hidden\n",
    "less .hidden/.combined.txt\n",
    "can use the -a (show all) switch to ls to make it show everything in a directory, including the hidden files and folders:\n",
    "ls\n",
    "ls -a\n",
    "ls .hidden\n",
    "ls -a .hidden\n",
    "As for our recently installed tree command, that works in a similar way (except without an appearance by . and ..):\n",
    "tree\n",
    "tree -a\n",
    "rm -r /tmp/tutorial\n",
    "ls /tmp\n",
    "As a last step, let’s close the terminal. You can just close the window, but it’s better practice to log out of the shell. You can either use the logout command, or the Ctrl-D keyboard shortcut. If you plan to use the terminal a lot, memorising Ctrl-Alt-T to launch the terminal and Ctrl-D to close it will soon make it feel like a handy assistant that you can call on instantly, and dismiss just as easily\n",
    "\n",
    "<http://linuxcommand.org/tlcl.php>\n",
    "curl, wget, pip, npm, make,\n",
    "Evolution is the new software\n",
    "#To know the IP server dig +short myip.opendns.com @resolver1.opendns.com\n",
    "https://xie186.github.io/Novice2Expert4Bioinformatics/install-bioinformatics-software-in-linux.html\n",
    "https://hackr.io/blog/basic-linux-commands\n",
    "https://youtu.be/7-G_dQr7B44\n",
    "Install MUSCLE using binaries:\n",
    "https://youtu.be/UJx34KVLIgI\n",
    "Install PhyML using binaries:\n",
    "https://github.com/vappiah/bioinfoscr...\n",
    "Gene extractor script is available here\n",
    "https://itol.embl.de/\n",
    "iTOL browser\n",
    "https://youtu.be/2tMQYi_12IQ\n",
    "Python codes explained\n",
    "Installing muscle\n",
    "https://www.drive5.com/muscle/downloads.htm\n",
    "Select Linux 64x copy link\n",
    "Then control alt T to open the terminal\n",
    "To download use: wget “link”\n",
    "To execute: ar xvfz “filepath”\n",
    "Installing RAxML\n",
    "Copy the zip link\n",
    "CLIPKIT - trimming\n",
    "https://jlsteenwyk.com/ClipKIT/advanced/index.html\n",
    "https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.3001007\n",
    "Mesquite\n",
    "cd bin\n",
    "/bin$ cd MrBayes\n",
    "cd src ./mb\n",
    "\n",
    "#move a list of files into a new folder\n",
    "mv-t dest_dir $(< text.file)\n",
    "\n",
    "\n",
    "Estructura \n",
    "Command [options]\n",
    "Ejemplos:\n",
    "Ls /path/\n",
    "Pwd\n",
    "Cat concatenate\n",
    "Cd change directory\n",
    "• cut - Cut: Choose a column from a tab-delimited file (spreadsheet)\n",
    "• Example: cut -f (column number to choose) file.txt\n",
    "• head - Head: See the first few lines of a file\n",
    "• Example: head -n (number of lines to show) file.txt\n",
    "• ls - List: Show the content of a directory\n",
    "• Example: ls /directory/to/show\n",
    "• mkdir - Make directory: Make a new directory/folder\n",
    "• Example: mkdir name_of_new_directory\n",
    "• mv - Move: Move a file(s) or directory\n",
    "• Example: mv /file/at/old/location/ /new/location/to/move/to\n",
    "• rm - Remove: Delete a file/directory (must use -r option to delete a directory)\n",
    "• Example: rm name_of_file_to_delete\n",
    "#To download \n",
    "wget http://address\n",
    "-o redirects to path\n",
    "-o- redirects to standard output for piping\n",
    "-q supresses messages\n",
    "\n",
    "From <https://cyverse-leptin-rna-seq-lesson-dev.readthedocs-hosted.com/en/latest/section-3.html> \n",
    "\n",
    "special characters (e.g. !#$%^&* etc.). Also, you cannot use spaces in file names\n",
    "\n",
    "From <https://cyverse-leptin-rna-seq-lesson-dev.readthedocs-hosted.com/en/latest/section-3.html> \n",
    "\n",
    "https://cyverse-leptin-rna-seq-lesson-dev.readthedocs-hosted.com/en/latest/section-3.html\n",
    "https://www.freecodecamp.org/espanol/news/grep-command-tutorial-how-to-search-for-a-file-in-linux-and-unix-with-recursive-find/\n",
    "https://tldp.org/LDP/Bash-Beginners-Guide/html/sect_09_02.html\n",
    "https://datacarpentry.org/wrangling-genomics/\n",
    "https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-with-commands\n",
    "https://sanbomics.com/2022/01/08/complete-rnaseq-alignment-guide-from-fastq-to-count-table/\n",
    "RNAseq tutorial - part 1 - building STAR genome index\n",
    "Virus genome assembly and consensus sequence from reads\n",
    "https://cyverse-leptin-rna-seq-lesson-dev.readthedocs-hosted.com/en/latest/section-8.html\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "https://gist.github.com/darencard/e1933e544c9c96234e86d8cbccc709e0\n",
    "\n",
    "https://www.thegeekstuff.com/2010/03/awk-arrays-explained-with-5-practical-examples/\n",
    "https://notebook.community/crazyhottommy/scripts-general-use/Shell/Awk_anotates_vcf_with_bed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "whoami, pwd, ls, ls -l\n",
    "\n",
    "\n",
    "sudo apt install openjdk-17-jdk\n",
    "https://gist.github.com/DannyArends/04d87f5590090dfe0dc6b42e5e1bbe15\n",
    "RNA Sequencing - Setup and Prerequisites\n",
    "\n",
    "https://www.youtube.com/watch?v=PlqDQBl22DI&t=14s\n",
    "\n",
    "How to manipulate gene expression data from NCBI GEO in R using dplyr | Bioinformatics for beginners\n",
    "\n",
    "\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE19080\n",
    "\n",
    "\n",
    "\n",
    "Stem-like exhausted and memory CD8+ T cells in cancer\n",
    "In vivo clinical molecular imaging of T cell activity\n",
    "\n",
    "\n",
    "Sudo make install\n",
    "\n",
    "export PATH=$PATH:/home/dev105/devapps/STAR/source\n",
    "\n",
    "\n",
    "Cat  | xargs Rscript --vanilla pipeline.R \n",
    "\n",
    "#Read the sample from the command line\n",
    "sranumber<-commandArgs(trailingOnly=TRUE) \n",
    "print(sranumber)\n",
    "#Function Idk yet\n",
    "execute<-function(x, outputfile=NA, intern=FALSE, quitOnError=FALSE){\n",
    "  if(!is.na(outputfile) &&file.exists(outputfile)){\n",
    "    cat(\"Output for step exists, skipping this step\\n\");\n",
    "    invisible(\"\")\n",
    "  }\n",
    "  cat(\"----\", x, \"\\n\"); res<-system(x, intern=intern); cat(\">>>>\", res[1], \"\\n\")\n",
    "  if(res[1] >=1){ \n",
    "    cat(\"Error external process did not finish\\n\\n\");\n",
    "    if(quitOnError) q(\"no\")\n",
    "  }\n",
    "}\n",
    "\n",
    "input.dir<-\"/home/dev105/projects/rawdata\"\n",
    "input.base<-sranumber[1]\n",
    "output.dir<-paste0(\"/home/dev105/projects/outputdata/\", input.base,\".aln\")\n",
    "\n",
    "#Create an output folder\n",
    "if(!file.exists(input.dir)){ dir.create(input.dir, recursive=TRUE) }\n",
    "if(!file.exists(output.dir)){ dir.create(output.dir, recursive=TRUE) }\n",
    "\n",
    "#STEP 0 - SRA Download and Compress\n",
    "setwd(input.dir)\n",
    "execute(paste0(\"prefetch -p\", input.base)\n",
    "execute(paste0(\"fasterq-dump -p --split-files \", input.base), paste0(input.base, \"_1.fastq\"))\n",
    "execute(paste0(\"bgzip \", input.base, \"_1.fastq\"), paste0(input.base, \"_1.fastq.gz\"))\n",
    "execute(paste0(\"bgzip \", input.base, \"_2.fastq\"), paste0(input.base, \"_2.fastq.gz\"))\n",
    "\n",
    "#STEP 1 - READ Trimming\n",
    "trim.files<-c(\n",
    "                  paste0(input.dir, \"/\", input.base,\"_1.fastq.gz\"),\n",
    "                  paste0(input.dir, \"/\", input.base,\"_2.fastq.gz\"),\n",
    "                  paste0(output.dir, \"/\", input.base,\"_1.P.fastq.gz\"),\n",
    "                  paste0(output.dir, \"/\", input.base,\"_1.U.fastq.gz\"),\n",
    "                  paste0(output.dir, \"/\", input.base,\"_2.P.fastq.gz\"),\n",
    "                  paste0(output.dir, \"/\", input.base,\"_2.U.fastq.gz\")\n",
    "                )\n",
    "trim.path<-\"/home/dev105/devapps/Trimmomatic-0.39\"\n",
    "trim.exec<-paste0(\"java -jar \", trim.path, \"/trimmomatic-0.39.jar\")\n",
    "trim.opts<-paste0(\"ILLUMINACLIP:\",trim.path,\"/adapters/TruSeq3-PE-2.fa:2:30:10\")\n",
    "trim.opts<-paste0(trim.opts, \"LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\")\n",
    "trim.cmd<-paste0(trim.exec, \"PE \", paste0(trim.files, collapse=\"\"), \"\", trim.opts)\n",
    "\n",
    "execute(trim.cmd, trim.files[3])\n",
    "\n",
    "q(\"no\")\n",
    "\n",
    "\n",
    "\n",
    "My index for genome star alignment\n",
    "https://labshare.cshl.edu/shares/gingeraslab/www-data/dobin/STAR/STARgenomes/Human/\n",
    "https://data.broadinstitute.org/Trinity/\n",
    "https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
