{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sources\n",
    " * https://github.com/niemasd/ViralConsensus\n",
    " * https://hpc.nih.gov/apps/trimAl.html\n",
    " * http://www.iqtree.org/doc/Dating\n",
    " * https://en.wikipedia.org/wiki/Ancestral_reconstruction\n",
    " * https://bootstrappers.umassmed.edu/bootstrappers-courses/courses/rCourse/\n",
    " * https://bookdown.org/pdr_higgins/rmrwr/\n",
    " * https://rmarkdown.rstudio.com/lesson-2.html\n",
    " * https://www.markdownguide.org/cheat-sheet/\n",
    " * https://learnmetabarcoding.github.io/LearnMetabarcoding/gettingstarted/cli_bioinformatics/cheatsheet.html\n",
    " * https://evol.bio.lmu.de/_statgen/compevol/PhyloHandout.pdf\n",
    " * https://viralzone.expasy.org/5576\n",
    " * https://www.biostars.org/p/418165/\n",
    " * https://github.com/inab/trimal\n",
    " * https://jlsteenwyk.com/ClipKIT/other_software/index.html\n",
    " * https://phyluce.readthedocs.io/en/latest/daily-use/daily-use-3-uce-processing.html\n",
    " * https://astrobiomike.github.io/genomics/phylogenomics\n",
    " * https://github.com/mmatschiner/tutorials\n",
    " * https://yulab-smu.top/treedata-book/index.html\n",
    " * https://jvanheld.github.io/using_IFB_NNCR/practicals/blast_proteome/blast_proteome.html\n",
    "https://treetime.readthedocs.io/en/latest/tutorials/timetree.html\n",
    "https://academic.oup.com/ve/article/4/1/vex042/4794731?login=true\n",
    "https://github.com/neherlab/treetime_examples/blob/master/data/ebola/ebola.metadata.csv\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5860581/\n",
    "https://hal.science/hal-01334934/document\n",
    "https://taming-the-beast.org/tutorials/FBD-tutorial/FBD-tutorial.pdf\n",
    "https://cran.r-project.org/web/packages/Rphylopars/Rphylopars.pdf\n",
    "https://yulab-smu.top/treedata-book/chapter1.html\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6294524/\n",
    "http://www.iqtree.org/doc/iqtree-doc.pdf\n",
    "https://hal.science/hal-01334934/document\n",
    "https://en.wikipedia.org/wiki/Ancestral_reconstruction\n",
    "\n",
    "* https://treetime.readthedocs.io/en/latest/tutorials/timetree.html\n",
    "https://academic.oup.com/ve/article/4/1/vex042/4794731?login=true\n",
    "https://github.com/neherlab/treetime_examples/blob/master/data/ebola/ebola.metadata.csv\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5860581/\n",
    "https://hal.science/hal-01334934/document\n",
    "https://taming-the-beast.org/tutorials/FBD-tutorial/FBD-tutorial.pdf\n",
    "https://cran.r-project.org/web/packages/Rphylopars/Rphylopars.pdf\n",
    "https://yulab-smu.top/treedata-book/chapter1.html\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6294524/\n",
    "http://www.iqtree.org/doc/iqtree-doc.pdf\n",
    "https://hal.science/hal-01334934/document\n",
    "  * https://en.wikipedia.org/wiki/Ancestral_reconstruction\n",
    "\n",
    "  * https://github.com/phac-nml/refseq_masher  #In case that you need to know what genomes are included in your sample \n",
    "\n",
    "    https://github.com/amkozlov/raxml-ng\n",
    "\n",
    "    https://hal.science/hal-02535311/document\n",
    "\n",
    "\n",
    "https://nbisweden.github.io/workshop-plotting-in-r/2109/lab_phylo.html\n",
    "https://yulab-smu.top/treedata-book/chapter5.html\n",
    "\n",
    "https://mafft.cbrc.jp/alignment/software/manual/manual.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "## make a list of accesion numbers\n",
    "seq -f \"MT113%g\" 123 153 > 2020liao.txt\n",
    "seq -f \"MT113%g\" 155 178 >> 2020liao.txt \n",
    "while read line;\n",
    "do\n",
    "$line;\n",
    "done <\n",
    "oartition 1-759 ()\n",
    "5187-6653\n",
    "5203..6669 (NCBI)\n",
    "DNA, part1 = 1-759\n",
    "DNA, part2 = 5187-6653\n",
    "gp46 = 5015-\n",
    "gp46 = 5187-6129\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "mamba create --name pipdev2 python=2.7\n",
    "mamba activate pipdev2\n",
    "#mamba deactivate pipdev2\n",
    "\n",
    "path=/media/dev105/HD-PCGU3-A/0305_ancient/repo/sra/fastq\n",
    "cd /media/dev105/HD-PCGU3-A/0305_ancient/repo/sra/fastq\n",
    "for i in $(ls -1 *pass_1.fastq | sed 's/\\_pass_1.fastq//'); do\n",
    "path=/media/dev105/HD-PCGU3-A/0305_ancient/repo/sra/fastq\n",
    "reference=/home/dev105/OneDrive/phyloseq0919/data/refgenome_HTLV/J02029.1.fasta\n",
    "output=~/projects\n",
    "echo $i\n",
    "minimap2 -t 5 -ax sr $reference $path/${i}_pass_1.fastq $path/${i}_pass_2.fastq | viral_consensus -i - -r $reference --min_qual 10 --min_depth 1 --min_freq 0.1 -o $output/$i\\.fa; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "cd /media/dev105/HD-PCGU3-A/SRP042225/SRP042225\n",
    "#Output folder\n",
    "fol=/media/dev105/HD-PCGU3-A/2405_htlvgenome/2405_output\n",
    "#reference genome to remove\n",
    "refg=/media/dev105/HD-PCGU3-A/2405_htlvgenome/2405_ref/GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.gz\n",
    "#loop\n",
    "for i in $(ls -1 *_1.fastq.gz | sed 's/\\_1.fastq.gz//'); do \n",
    "    echo \"$i\" #print message\n",
    "    #first FASTP for quality\n",
    "    fastp -i ${i}\\_1.fastq.gz -I ${i}\\_2.fastq.gz -o $fol/${i}\\_R1.fastq.gz -O $fol/${i}\\_R2.fastq.gz\n",
    "    minimap2 --split-prefix=tmp$$ -a -xsr $refg $fol/${i}\\_R1.fastq.gz $fol/${i}\\_R2.fastq.gz | \n",
    "    samtools view -bh | \n",
    "    samtools sort -o $fol/${i}.bam\n",
    "    samtools index  $fol/${i}.bam\n",
    "    samtools fastq -F 3584 -f 77 $fol/${i}.bam  | gzip -c > $fol/${i}\\_nh\\_R1.fastq.gz\n",
    "    samtools fastq -F 3584 -f 141 $fol/${i}.bam | gzip -c > $fol/${i}\\_nh\\_R2.fastq.gz\n",
    "    bwa mem -t 5 /media/dev105/HD-PCGU3-A/2405_htlvgenome/2405_ref/J02029.1.fasta $fol/${i}\\_nh\\_R1.fastq.gz $fol/${i}\\_nh\\_R2.fastq.gz | \n",
    "    samtools --threads 5 sort  -o  $fol/${i}\\_htlv.bam\n",
    "    #third, pileup using Q20 and ivar -t 0.8\n",
    "    samtools mpileup -aa -A -d 1000000 -Q 20 $fol/${i}\\_htlv.bam | \n",
    "    ivar consensus -t .8 -m 10 -p  $fol/${i}\\_htlv.fa; done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "Raw FASTQ files were read using two approaches:\n",
    "# three folders were analized 181215_Hiseq3000  460 181225_Hiseq3000  689 190104_Hiseq3000\n",
    "#a folder with all reads R1 and R3 were reads\n",
    "# Mode 01 using BWA MEM + IVAR\n",
    "for i in $(ls -1 *R1_001.fastq.gz | sed 's/\\R1_001.fastq.gz//'); do \n",
    "    f1=Peru_$(echo $i | sed 's/.*Nakahata_//'| sed 's/_L00._//') #definition of f1\n",
    "    echo \"$i\"\"->\"\"$f1\" #print message\n",
    "    #first FASTP for quality\n",
    "    fastp -i ${i}\\R1_001.fastq.gz -I ${i}\\R3_001.fastq.gz -o ~/projects/seq0914/${f1}\\_R1.fastq.gz -O ~/projects/seq0914/${f1}\\_R2.fastq.gz\n",
    "    #second obtain the bam file\n",
    "    bwa mem -t 5 ~/projects/rawseq0824/J2029/J02029.1.fasta ~/projects/seq0914/${f1}\\_R1.fastq.gz ~/projects/seq0914/${f1}\\_R2.fastq.gz | \n",
    "    samtools --threads 5 sort  -o ~/projects/seq0914/$f1\\.bam\n",
    "    #third, pileup using Q20 and ivar -t 0.8\n",
    "    samtools mpileup -aa -A -d 1000000 -Q 20 ~/projects/seq0914/$f1\\.bam | \n",
    "    ivar consensus -t .8 -m 10 -p  ~/projects/seq0914/fasta/$f1\\.fa\n",
    "    rm ~/projects/seq0915/$f1\\.bam ~/projects/seq0915/${f1}\\_R1.fastq.gz ~/projects/seq0915/${f1}\\_R2.fastq.gz; done\n",
    "\n",
    "\n",
    "#Mode 02 using minimap2 + viral_consensus\n",
    "for i in $(ls -1 *R1_001.fastq.gz | sed 's/\\R1_001.fastq.gz//'); do \n",
    "    f1=Peru_$(echo $i | sed 's/.*Nakahata_//'| sed 's/_L00._//')\n",
    "    echo \"$i\"\"->\"\"$f1\"\n",
    "    #first FASTP for quality    \n",
    "    fastp -i ${i}\\R1_001.fastq.gz -I ${i}\\R3_001.fastq.gz -o ~/projects/seq0915/${f1}\\_R1.fastq.gz -O ~/projects/seq0915/${f1}\\_R2.fastq.gz\n",
    "    minimap2 -t 5 -ax sr ~/projects/rawseq0824/J2029/J02029.1.fasta ~/projects/seq0915/${f1}\\_R1.fastq.gz  ~/projects/seq0915/${f1}\\_R2.fastq.gz |\n",
    "    viral_consensus -i - -r ~/projects/rawseq0824/J2029/J02029.1.fasta -o ~/projects/seq0915/fasta/$f1\\.fa; done\n",
    "\n",
    "#Mode 02 but with t 0.8 for next project about comparison\n",
    "for i in $(ls -1 *R1_001.fastq.gz | sed 's/\\R1_001.fastq.gz//'); do \n",
    "f1=Peru_$(echo $i | sed 's/.*Nakahata_//'| sed 's/_L00._//')\n",
    "path=/home/dev105/projects/seq0917\n",
    "echo \"$i\"\"->\"\"$f1\"\n",
    "fastp -i ${i}\\R1_001.fastq.gz -I ${i}\\R3_001.fastq.gz -o $path/${f1}\\_R1.fastq.gz -O $path/${f1}\\_R2.fastq.gz\n",
    "bwa mem -t 5 ~/projects/rawseq0824/J2029/J02029.1.fasta $path/${f1}\\_R1.fastq.gz $path/${f1}\\_R2.fastq.gz | samtools sort  -o $path/$f1\\.bam\n",
    "samtools mpileup -aa -A -d 1000000 -Q 20 $path/$f1\\.bam | ivar consensus -t .8 -m 10 -p  $path/fasta_ivar/$f1\\.fa\n",
    "minimap2 -t 5 -ax sr ~/projects/rawseq0824/J2029/J02029.1.fasta $path/${f1}\\_R1.fastq.gz  $path/${f1}\\_R2.fastq.gz | viral_consensus -i - -f 0.8 -r ~/projects/rawseq0824/J2029/J02029.1.fasta -o $path/fasta_vircons/$f1\\.fa\n",
    "rm $path/$f1\\.bam $path/${f1}\\_R1.fastq.gz  $path/${f1}\\_R2.fastq.gz ; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#Manipulating sequences\n",
    "ls -ltr | awk '/Japan.*fa/ {print \"mv \"$NF\" \"substr($NF,1,15)}'\n",
    "#to rename files\n",
    "for i in $(ls GCF*.gz); do var=$(echo $i | awk -F \"_\" '{print $1\"_\"$2}'); echo $var; mv $i $var; done\n",
    "for i in $(ls *.fa); do var=$(echo $i | awk -F \"_\" '{print \"Japan_\" $3 \"_\" $4}'); echo $var; mv $i $var; done\n",
    "for i in $(ls *.fa); do var=$(echo $i | awk -F \".\" '{print $1 \".fa\"}'); echo $var; mv $i $var; done\n",
    "#rename other folder\n",
    "ls Peru_*_IRID*.fa\n",
    "for i in $(ls Peru_*_IRID*.fa); do var=$(echo $i | awk -F \"_\" '{print $1 \"_\" $3 \"_\" $4}'); echo $var; mv $i $var; done\n",
    "#rename the header with file names\n",
    "for file in *.fa;    do        sed -i \"s/>.*/${file%%.*}/\" \"$file\" ; done\n",
    "sed -i \"s/Japan.*/^>./\" \n",
    "sed 's/Japan/\\>Japan\\_vircons/g'  japan0919vir_cons > japanvircons\n",
    "#Searching for gp46 in blast (ANOTHER PROJECT)\n",
    "#Some of the accessions provided ['AB273635.1', 'AJ493586.1', 'AJ493585.1', 'AJ493584.1', 'AJ493594.1', 'AJ493589.1', 'AJ493588.1', 'AJ493592.1', '70.74', 'M64607.1'] were not found\n",
    "datasets download virus genome accession --inputfile gp46_blast_0928.txt\n",
    "#delete repeated headers\n",
    "awk '/^>/ { f = !a[$0]++ } f' all_seq_maffted > all_seq_maffted_uniq\n",
    "#to rename multiple files in a folder txt -> pdf\n",
    "for a in *.txt; do mv -- \"$a\" \"${a%.txt}.pdf\"; done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "\n",
    "#Get genomes in Fasta and metadata from a list\n",
    "datasets download virus genome accession --inputfile list.txt\n",
    "dataformat tsv virus-genome --package /home/dev105/OneDrive/phyloseq0919/data/seq_raw/ncbi_seq/Raw_HTLV_1only/ncbi_dataset.zip > dataformat.tsv\n",
    "\n",
    "#filtering \n",
    "cd /home/dev105/OneDrive/phyloseq0919/data/seq_raw/ncbi_seq/Raw_HTLV_1only\n",
    "cat list328fastncbi.fa | seqkit grep -r -p M86840.1 -p LC378575.1 -p L03562.2 -p L03561.2 -p D13784.1 -p AB979451.1 -v > seqkit_ncbifilteredseq_1124.fa\n",
    "#to check\n",
    "grep -A1 -i M86840.1 seqkit_ncbifilteredseq_1124.fa\n",
    "\n",
    "#remove descriptions after space\n",
    "awk -F \"|\" '/^>/ {print $1;next}1' file.fa > simplenames.fa\n",
    "awk sub{/.*Nakahata_/,\"Peru\",\"_\",$NF;}1'\n",
    "#remove all letter after space\n",
    "sed '/^>/ s/ .*//'\n",
    "#to add the >:\n",
    "sed 's/Peru/\\>Peru/g' cgperu1.fa >cgperu2.fa\n",
    "seqkit replace -p \"\\s.+\" seqkit_ncbifilteredseq_1124.fa > seqkit_ncbifiltered_nodesc_1124.fasta\n",
    "\n",
    "#remove duplicates (no duplicates in this case)\n",
    "seqkit rmdup -s seqkit_ncbifiltered_nodesc_1124.fasta > seqkit_wodup_filtered_1124.fa -d otro.fa\n",
    "\n",
    "#remove N on the sequences (not necessay yet)\n",
    "awk '!/^>/{gsub(/[^CTAG]/, \"-\")}1'\n",
    "seqkit seq -g seqkit_ncbifiltered_nodesc_1124.fasta > seqkit_nfilt_nodesc_nonn_11.24.fasta\n",
    "\n",
    "# count number of sequences in original peru file\n",
    "grep \">\" file.fasta | wc -l\n",
    "\n",
    "# change the names (erase Peru_ or Japan_)\n",
    "sed 's/\\>Peru_/\\>/g'\n",
    "sed 's/>Peru_ivar_/>/g' totalivar.fa > peru_shortfasta.fa\n",
    "sed 's/>Japan_ivar_/>JP_/g' peru_shortfasta.fa >shortfasta.fa\n",
    "sed 's/>SS/>PE_SS/g' shortfasta.fa >shortfinalunfiltered.fa\n",
    "\n",
    "# replace NN by - (noyet)\n",
    "awk '!/^>/{gsub(/[^CTAG]/, \"-\")}1' shortfinal.fasta > shortwithoutN.fa\n",
    "#filter sequences by name (less than 6000)\n",
    "cat | seqkit grep -r -p IRID011_S130 -p IRID038_S122 -p IRID040_S124 -p IRID067_S233 -p JP_14_S84 -p JP_17_S87 -v > seqkit_perufiltered_1124.fa\n",
    "#remove duplicates by sequence\n",
    "seqkit rmdup -s seqkit_perufiltered_1124.fa > seqkit_wodup_filtered_1124.fa -d otro.fa\n",
    "#create a preliminary list\n",
    "grep -e \">\" seqkit_wodup_filtered_1124.fa |wc -l\n",
    "# seqkit_ncbifiltered_nodesc_1124.fasta (final file with N)\n",
    "\n",
    "#join two files /home/dev105/OneDrive/phyloseq0919/data/mature_data\n",
    "#cat seqkit_nfilt_nodesc_nonn_11.24.fasta seqkit_wodup_filtered_1124.fa >final_filtered_woduplicates_1124.fa\n",
    "cd /home/dev105/OneDrive/phyloseq0919/data/mature_data\n",
    "mafft --6merpair --thread -4 --keeplength  final_filtered_woduplicates_1124.fa /home/dev105/OneDrive/phyloseq0919/refgenome_HTLV/J2029/J02029.1.fasta\n",
    " > final_fil1_dup0_1124.fa\n",
    "\n",
    "\n",
    " #aalignment\n",
    "mafft --6merpair --localpair --maxiterate 1000 --thread 4 --keeplength --addfragments final_filtered_woduplicates_1124.fasta /home/dev105/OneDrive/phyloseq0919/refgenome_HTLV/J2029/J02029.1.fasta > final\n",
    "# trimming manually 22 and 13 nucleotides at both ends (aliview)\n",
    "/home/dev105/OneDrive/phyloseq0919/data/mature_data/final_mafft1_filt1_wodup1_gap1_1124.fasta\n",
    "#new duplicates appeared\n",
    "seqkit rmdup -s final_mafft1_filt1_wodup1_gap1_1124.fasta > seqkit_mafft1_filt1_wodup1_gap1_dup0_1124.fasta -d duplicates.fa\n",
    "grep \">\" duplicates.fa\n",
    "#395 sequences remain\n",
    "grep -e \">\" seqkit_mafft1_filt1_wodup1_gap1_dup0_1124.fasta | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "cat maffted/seqkit_mafft1_filt1_wodup1_gap1_dup0_1124.fasta | seqkit grep -r -p KF797883.1 -p KY007245.1 -p KY007246.1  -v > seqkit_norecomb_1201.fasta\n",
    "remains 392\n",
    "#an additional list for simplot y GARD\n",
    "cat maffted/seqkit_mafft1_filt1_wodup1_gap1_dup0_1124.fasta | seqkit grep -r -p KF797883.1 -p KY007245.1 -p KY007246.1 -p KF797833.1 -p KY007272.1 -p KY007262.1 -p KY007253.1 -p KX905202.1 -p L02534.1 > recombinant_list_1201.fasta\n",
    "\n",
    "raxml-ng --all --threads 5 --bs-trees 1000 --msa-format FASTA --data-type DNA --prefix raxml_1124 --msa  seqkit_norecomb_1201.fasta --model  partition.txt --bs-cutoff 0.03 --blmin 1e-5 --outgroup \"L02534.1\",\"KX905203.1\", \"KX905202.1\", \"KF242506.1\", \"KF242505.1\", \"JX891479.1\", \"JX891478.1\"\n",
    "seqkit_norecomb_1201\n",
    "\n",
    "#Annotation\n",
    "prokka -kingdom Viruses --outdir htlv_prokka --genus Deltaretrovirus J02029.1.fasta\n",
    "\n",
    "raxml-ng --all --threads 5 --bs-trees 10000 --msa-format FASTA --data-type DNA --prefix raxml_1124 --msa  seqkit_mafft1_filt1_wodup1_gap1_dup0_1124.fasta --model  partition.txt --bs-cutoff 0.03 --blmin 1e-5 --outgroup \"L02534.1\",\"KX905203.1\", \"KX905202.1\", \"KF242506.1\", \"KF242505.1\", \"JX891479.1\", \"JX891478.1\"\n",
    "\n",
    "\n",
    "raxml-ng --bootstrap --bs-trees 1000 --msa seqkit_norecom_321_1207.fasta --model partition.txt --prefix ~/OneDrive/phyloseq0919/results/rxaml/1208_rxaml_2 --seed 5 --threads 5\n",
    "\n",
    " raxml-ng --bsconverge --bs-trees raxml_1124.raxml.bootstraps --prefix converge --seed 2 --\n",
    "threads 1 --bs-cutoff 0.01\n",
    " converge after 600\n",
    "\n",
    " raxml-ng --support --tree raxml_1124.raxml.bestTree --bs-trees raxml_1124.raxml.bootstraps --prefix 1208_support --threads 1 --seed 2 --bs-metric fbp, tbe\n",
    "\n",
    "\n",
    "raxml-ng --all --threads 5 --bs-trees 1000 --msa-format FASTA --data-type DNA --prefix raxml_1124 --msa  seqkit_norecomb_1201.fasta --model  partition.txt --bs-cutoff 0.03 --blmin 1e-5 \n",
    "\n",
    "\n",
    "#First necessary to find the right model\n",
    "#nexus\n",
    "begin sets;\n",
    "\tcharset 5LTR_3LTR = 1-755  8279-9033;\n",
    "\tcharset env_gp46_gp21 = 5181-6647\\3;\n",
    "\tcharset restHTLV = 756-5180 6648-8278;\n",
    "end;\n",
    "\n",
    "raxml-ng --all --threads 5 --bs-trees 10000 --msa-format FASTA --data-type DNA --prefix raxml_1124 --msa  seqkit_mafft1_filt1_wodup1_gap1_dup0_1124.fasta --model bestmodel_101323_CG_allpartitions.best_scheme.nex --bs-cutoff 0.03 --blmin 1e-5 \n",
    "\n",
    "#################Run 01-16 All sequences but without JP\n",
    "raxml-ng --all --threads 6 --bs-trees 10000 --msa-format FASTA --data-type DNA --prefix raxml_0116 --msa seqkit_norec_noJP_0112.fasta --model partition.txt --bs-cutoff 0.03 --blmin 1e-5\n",
    "###Coverge evaluation\n",
    "raxml-ng --bsconverge --bs-trees raxml_0116.raxml.bootstraps --prefix bsconverge --seed 2 --threads 5 --bs-cutoff 0.01\n",
    "#########new run\n",
    "raxml-ng --all --threads 6 --bs-trees 10000 --msa-format FASTA --data-type DNA --prefix raxml_0216 --msa seqkit_norec_noJP_0112.fasta --model partition.txt --bs-cutoff 0.01 --blmin 1e-3\n",
    "\n",
    "#Another option is to use IQTREE\n",
    "iqtree2 -s  seqkit_mafft1_filt1_wodup1_gap1_dup0_1124.fasta  -p nexus.nex -m MF -mtree -T 5 --prefix MF_\n",
    "iqtree2 -s 0919maffted_allcases_trimmed-manual281_v2_092223.fa -p part.nex --polytomy  -blmin 1e-6 -B 10000 --prefix parti_ -T 4\n",
    "iqtree2 -s 0919maffted_allcases_trimmed-manual281_v2_092223.fa -p part.nex --polytomy  -blmin 3.3e-7 -B 10000 --prefix parti_ -T 4\n",
    "find the best model for gp46\n",
    "iqtree2 -s (sequences o dir) -p (partition) -m (modelo) -B 1000 -alrt 1000 -T 5\n",
    "m MF+MERGE find the best partition scheme\n",
    "--sampling GENESITE\n",
    "iqtree2 -s 0923gp46all_2020liao.fa -m TN+F+I+R2 -B 1000 -alrt 1000 --polytomy  -blmin 3.3e-7 -T 5  -prefix gp46tree_\n",
    "iqtree2 -s 0923gp46all_2020liao.fa -m TN+F+I+R2 -B 10000 -alrt 10000 --polytomy  -blmin 3.3e-7 -T 5  --prefix gp46tree\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
