{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Graphics\n",
    "\n",
    "\n",
    "https://github.com/hally166/Cytometry-R-scripts/blob/master/flowWorkspace.R\n",
    "\n",
    "(3) Flow Cytometry Data Analysis in R: gating with flowWorkspace\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "https://ib.berkeley.edu/courses/ib200/2018/labs/06/lab06.pdf\n",
    "https://academic.oup.com/sysbio/article/65/4/726/1753608\n",
    "https://pubmed.ncbi.nlm.nih.gov/37376536/\n",
    "https://pubmed.ncbi.nlm.nih.gov/36325034/\n",
    "https://taming-the-beast.org/workshops/\n",
    "\n",
    "https://epirhandbook.com/en/descriptive-tables.html\n",
    "https://geobgu.xyz/r/tables-conditionals-and-loops.html\n",
    "https://scriptsandstatistics.wordpress.com/2021/11/01/how-to-create-a-descriptive-summary-table-table-1-using-r/\n",
    "https://bookdown.org/kdonovan125/ibis_data_analysis_r4/objects-and-functions.html\n",
    "Inkscape seems a good alternative\n",
    "https://inkscape.org/doc/tutorials/basic/tutorial-basic.html\n",
    "\tØ Other can be' https://www.gimp.org/\n",
    "https://bioicons.com/\n",
    "https://github.com/duerrsimon/bioicons\n",
    "https://smart.servier.com/\n",
    "\n",
    "\n",
    "https://www.tidyverse.org/blog/2021/02/svglite-2-0-0/\n",
    "\n",
    "https://jchellmuth.com/posts/FACS-with-R/\n",
    "https://www.bioconductor.org/packages/release/bioc/html/flowCHIC.html\n",
    "https://github.com/kordastilab/ImmunoCluster\n",
    "https://dillonhammill.github.io/CytoExploreR/\n",
    "\n",
    "\n",
    "\n",
    "https://r-graph-gallery.com/circular-barplot.html\n",
    "https://bioinformatics.ccr.cancer.gov/docs/data-visualization-with-r/Lesson2_intro_to_ggplot/\n",
    "https://technicaljockey.com/r-programming-training/ggpubr-package-in-r/\n",
    "https://jtr13.github.io/cc20/index.html\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://danhdtruong.com/Making-a-professional-website/\n",
    "\n",
    "https://www.michaelgerth.net/news--blog/how-to-efficiently-bulk-download-ngs-data-from-sequence-read-databases\n",
    "\n",
    "https://www.ibm.com/support/fixcentral/swg/downloadOptions\n",
    "https://danhdtruong.com/SRA-conversion/ \n",
    "\n",
    "\n",
    "https://www.virtualbox.org/wiki/Downloads\n",
    "#07-13 Download virtualbox\n",
    "\n",
    "Skip unattended installation (because it make only to use as guest)\n",
    "\n",
    "#how to search SRA links\n",
    "https://sra-explorer.info/\n",
    "\n",
    "Developed by https://github.com/ewels/sra-explorer\n",
    "\n",
    "Alternatives:\n",
    "\n",
    "https://github.com/pachterlab/ffq\n",
    "\n",
    "\n",
    "Aspera configuration\n",
    "https://www.biostars.org/p/325010/\n",
    "https://github.com/aertslab/install_aspera_connect\n",
    "\n",
    "Script\n",
    "https://github.com/gerthmicha/perlscripts/blob/master/sra_download.pl\n",
    "\n",
    "https://rnnh.github.io/bioinfo-notebook/docs/fastq-dump_to_featureCounts.html\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "https://learnmetabarcoding.github.io/LearnMetabarcoding/gettingstarted/cli_bioinformatics/loops.html\n",
    "\n",
    "#RNA pipeline tutorial\n",
    "Setup RNA-Seq Pipeline from scratch: fastq (reads) to counts | Step-by-Step Tutorial\n",
    "#script\n",
    "https://github.com/kpatel427/YouTubeTutorials/blob/main/RNASeqpipeline.sh\n",
    "\n",
    "\n",
    "\n",
    "https://bioinformaticsdotca.github.io/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://sdu-kz.libanswers.com/faq/247832\n",
    "\n",
    "https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0248001\n",
    "\n",
    "Equipo, por favor\n",
    "\n",
    "https://1drv.ms/x/s!AgLeOtW-pHFPh6c16AnvMZqx0i-uqA\n",
    "\n",
    "https://www.holiday-sc.jp/sp/shops/094-kamoike/?shop=094-kamoike\n",
    "\n",
    "https://joyfit.jp/masagohonmachi/\n",
    "\n",
    "https://www.nippon.com/en/japan-data/h01397/\n",
    "\n",
    "https://pubmed.ncbi.nlm.nih.gov/36791318/\n",
    "\n",
    "https://pubmed.ncbi.nlm.nih.gov/25875653/\n",
    "\n",
    "https://jhudatascience.org/tidyversecourse/get-data.html\n",
    "\n",
    "https://yulab-smu.top/treedata-book/chapter2.html#fig:reroot\n",
    "\n",
    "https://yulab-smu.top/treedata-book/chapter5.html\n",
    "\n",
    "https://bioc.ism.ac.jp/packages/3.5/bioc/vignettes/ggtree/inst/doc/treeAnnotation.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "https://github.com/bruvellu/sra.py/blob/master/fetch_sra.py\n",
    "\n",
    "https://colab.research.google.com/github/saketkc/pysradb/blob/master/notebooks/09.Query_Search.ipynb\n",
    "\n",
    "https://epirhandbook.com/en/diagrams-and-charts.html#flow-diagrams\n",
    "https://github.com/paulstothard/ncbi_search\n",
    "\n",
    "# Introduction to Shell and Cluster computing (HMS-RC’s O2 cluster) | Introduction to the command line interface (Shell)\n",
    "\n",
    "https://hbctraining.github.io/Intro-to-shell-flipped/schedule/links-to-lessons.html\n",
    "\n",
    "## Learning Objectives\n",
    "\n",
    "- Understand the necessity for, and use of, the command line interface (bash/shell).\n",
    "\n",
    "## Installations\n",
    "\n",
    "## Notes\n",
    "\n",
    "- These materials focus on the use of local computational resources at Harvard, which are **only accessible to Harvard affiliates**\n",
    "- Non-Harvard folks can [download the data](https://www.dropbox.com/s/3lua2h1oo18gbug/unix_lesson.tar.gz?dl=1) and set up to work on their local clusters (with the help of local system administrators)\n",
    "\n",
    "### Instructions for Harvard researchers with access to HMS-RC’s O2 cluster\n",
    "\n",
    "To run through the code in the lessons below, you will need to be **logged into O2** and **working on a compute node** (i.e. your command prompt should have the word `compute` in it).\n",
    "\n",
    "1. Log in using `ssh ecommonsID@o2.hms.harvard.edu` and enter your password.\n",
    "2. Once you are on the login node, use `srun --pty -p interactive -t 0-2:30 --mem 1G /bin/bash` to get on a compute node or as specified in the lesson.\n",
    "3. Proceed only once your command prompt has the word `compute` in it.\n",
    "4. If you log out between lessons (using the `exit` command twice), please follow points 1. and 2. above to log back in and get on a compute node when you restart with the self learning.\n",
    "\n",
    "## Lessons\n",
    "\n",
    "### Part I\n",
    "\n",
    "### Part II\n",
    "\n",
    "### Part III\n",
    "\n",
    "## Building on this workshop\n",
    "\n",
    "## Resources\n",
    "\n",
    "Cheat sheets:\n",
    "\n",
    "Online tutorials:\n",
    "\n",
    "General help:\n",
    "\n",
    "- Google it! - if you don’t know how to do something, try Googling it, other people have probably had the same question.\n",
    "- Learn by doing! There’s no real other way to learn this than by trying it out.\n",
    "- Use `man bash` to get more information about bash (bourne-again shell)\n",
    "\n",
    "*These materials have been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*\n",
    "\n",
    "[Workshop Schedule | Introduction to RNA-Seq using high-performance computing - ARCHIVED](https://www.notion.so/Workshop-Schedule-Introduction-to-RNA-Seq-using-high-performance-computing-ARCHIVED-c05386b722c442b79d9d21a48ccad32c?pvs=21)\n",
    "\n",
    "[GitHub - hbctraining/Intro-to-rnaseq-hpc-O2: This repository has teaching materials for a 2 and 3-day Introduction to RNA-sequencing data analysis workshop using the O2 Cluster](https://www.notion.so/GitHub-hbctraining-Intro-to-rnaseq-hpc-O2-This-repository-has-teaching-materials-for-a-2-and-3-da-e12a6b2300be471081d9c96fe6f3b45f?pvs=21)\n",
    "\n",
    "[Tutorial of RNA-seq tumor immunity analysis](https://www.notion.so/Tutorial-of-RNA-seq-tumor-immunity-analysis-db54e63ea9c14ba38c0d9f7093fc2532?pvs=21)\n",
    "\n",
    "[Introduction to Bayesian phylogenetics using MCMC - YouTube](https://www.notion.so/Introduction-to-Bayesian-phylogenetics-using-MCMC-YouTube-1e66828249ea4b4ea1b9cec687f8a29e?pvs=21)\n",
    "\n",
    "[liaoherui/VirStrain: An RNA virus strain-level identification tool for short reads.](https://www.notion.so/liaoherui-VirStrain-An-RNA-virus-strain-level-identification-tool-for-short-reads-f72fbb7a9bd146f8b3073d585306d790?pvs=21)\n",
    "\n",
    "[MetaPop: a pipeline for macro- and microdiversity analyses and visualization of microbial and viral metagenome-derived populations | Microbiome | Full Text](https://www.notion.so/MetaPop-a-pipeline-for-macro-and-microdiversity-analyses-and-visualization-of-microbial-and-viral--f6b9cc76292f457abf24c85276b5a0f4?pvs=21)\n",
    "\n",
    "[How to Run Auspice — Auspice documentation](https://www.notion.so/How-to-Run-Auspice-Auspice-documentation-1b8a9e96020e490daafcb94a3bdd11eb?pvs=21)\n",
    "\n",
    "[SeqKit - Ultrafast FASTA/Q kit](https://www.notion.so/SeqKit-Ultrafast-FASTA-Q-kit-0cef05bf7ed8447188c6d71124bc44f8?pvs=21)\n",
    "\n",
    "[Tutorial of RNA-seq tumor immunity analysis](https://www.notion.so/Tutorial-of-RNA-seq-tumor-immunity-analysis-b5d34245dffe4e77a55466c8a6e3c7f5?pvs=21)\n",
    "\n",
    "[Installing ClusTRace | ClusTRace | University of Helsinki](https://www.notion.so/Installing-ClusTRace-ClusTRace-University-of-Helsinki-b85f2900df5847d6966c3df520d3a836?pvs=21)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`#to make files from a list of fasta`\n",
    "\n",
    "`perl -ne 'if (/^>(\\S+)/) { close OUT; open OUT, \">$1.fasta\" } print OUT' genome.fasta`\n",
    "\n",
    "[NC_001436.1](https://www.ncbi.nlm.nih.gov/nuccore/NC_001436.1/) reference sequence https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000863585.1/\n",
    "\n",
    "https://genomicsbootcamp.github.io/book/genotype-files-in-practice.html#genotype-files-in-practice\n",
    "\n",
    "https://www.cog-genomics.org/plink/1.9/resources#teach"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Update sudo\n",
    "\n",
    "```jsx\n",
    "sudo apt update\n",
    "sudo apt upgrade\n",
    "```\n",
    "\n",
    "#Installing dependencies\n",
    "\n",
    "```jsx\n",
    "sudo apt install dirmngr gnupg apt-transport-https ca-certificates software-properties-common\n",
    "```\n",
    "\n",
    "#authenticate package\n",
    "\n",
    "`gpg --keyserver keyserver.ubuntu.com --recv-keys 3F32EE77E331692F`\n",
    "\n",
    "$ sudo apt-key adv --keyserver [keyserver.ubuntu.com](http://keyserver.ubuntu.com/) --recv-keys `3F32EE77E331692F`\n",
    "\n",
    "#Add CRN repository\n",
    "\n",
    "sudo add-apt-repository 'deb https://cloud.r-project.org/bin/linux/ubuntu focal-cran40/'\n",
    "\n",
    "#R installation\n",
    "\n",
    "sudo apt install r-base\n",
    "\n",
    "**libicu66 package error**\n",
    "\n",
    "Download http://security.ubuntu.com/ubuntu/pool/main/i/icu/libicu66_66.1-2ubuntu2_amd64.deb\n",
    "\n",
    "$ cd ~/Downloads\n",
    "\n",
    "$ sudo dpkg -i libicu66_66.1-2ubuntu2_amd64.deb\n",
    "\n",
    "#Check the R version\n",
    "\n",
    "$ sudo -i R\n",
    "\n",
    "#Install Rstudio\n",
    "\n",
    "sudo dpkg -i rstudio-2022.02.1-461-amd64.deb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "#conda update conda\n",
    "\n",
    "#to see list of environments\n",
    "\n",
    "conda env list\n",
    "\n",
    "conda activate myenv\n",
    "\n",
    "conda deactivate\n",
    "\n",
    "conda create —name prengs\n",
    "\n",
    "conda remove --name myenv --all\n",
    "\n",
    "#to manage channels, see list of channels\n",
    "\n",
    "`conda config --show channels`\n",
    "\n",
    "`conda config --remove channels NOT_WANTED`\n",
    "\n",
    "`conda config --append channels CHANNEL_NAME`\n",
    "\n",
    "Example: `conda config --append channels conda-forge`\n",
    "\n",
    "`conda config --add channels conda-forge` (try to not use this)\n",
    "\n",
    "First install conda\n",
    "\n",
    "Download for Linux\n",
    "\n",
    "Then run: **[Cryptographic hash verification](https://docs.conda.io/projects/conda/en/latest/user-guide/install/download.html#id5)**\n",
    "\n",
    "`sha256sum filename`\n",
    "\n",
    "sha256sum ~/Downloads/Anaconda3-2022.10-Linux-x86_64.sh\n",
    "\n",
    "e7ecbccbc197ebd7e1f211c59df2e37bc6959d081f2235d387e08c9026666acd\n",
    "\n",
    "Install: `bash Anaconda-latest-Linux-x86_64.sh`\n",
    "\n",
    "In my case I used: bash ~/Downloads/Anaconda3-2022.10-Linux-x86_64.sh\n",
    "\n",
    "/home/dev105/anaconda3\n",
    "\n",
    "**Using with fish shell**\n",
    "\n",
    "`conda init fish`\n",
    "\n",
    "Update: `conda update conda`\n",
    "\n",
    "Uninstall: `rm -rf ~/miniconda`\n",
    "\n",
    "- OPTIONAL: Remove the following hidden file and folders that\n",
    "may have been created in the home directory:\n",
    "    - `.condarc` file\n",
    "    - `.conda` directory\n",
    "    - `.continuum` directory\n",
    "    \n",
    "    By running:\n",
    "    \n",
    "    `rm -rf ~/.condarc ~/.conda ~/.continuum`\n",
    "    \n",
    "\n",
    "Bioconda\n",
    "\n",
    "```\n",
    "conda config --add channels defaults\n",
    "conda config --add channels bioconda\n",
    "conda config --add channels conda-forge\n",
    "conda config --set channel_priority strict\n",
    "```\n",
    "\n",
    "conda create -n myenv1\n",
    "\n",
    "conda env list\n",
    "\n",
    "conda activate myenv1\n",
    "\n",
    "conda install samtools\n",
    "\n",
    "install —quiet *without asking for confirmation\n",
    "\n",
    "conda install -n myenv1 -c bioconda raxml\n",
    "\n",
    "/home/dev105/anaconda3/envs/myenv1\n",
    "\n",
    "conda install -c bioconda sra-tools\n",
    "\n",
    "myenv1\n",
    "\n",
    "\n",
    "\n",
    "Alingment\n",
    "\n",
    "https://github.com/bredelings/BAli-Phy\n",
    "\n",
    "java -jar /home/yoshiyuki/Downloads/tomighty-0.7.1.jar\n",
    "\n",
    "su\n",
    "\n",
    "Control D"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bioinformagician\n",
    "\n",
    "Download geo data sets\n",
    "\n",
    "Data normalized (FPKM) reading in expression matrix\n",
    "\n",
    "locate using gunzip (.csv.gz to .csv)\n",
    "\n",
    "Do not open it with excel.\n",
    "\n",
    "Steps to manipulate data:\n",
    "\n",
    "- Reading in expression matrix (FPKM)\n",
    "- Fetching metadata/clinical data using GEOquery R package\n",
    "- Useful functions\n",
    "\n",
    "getwd() #obtain the working directory\n",
    "\n",
    "setwd()# set the working directory\n",
    "\n",
    "#Install libraries\n",
    "\n",
    "install.packages(””)\n",
    "\n",
    "dplyr\n",
    "\n",
    "tidyverse\n",
    "\n",
    "GEOquery\n",
    "\n",
    "#load libraries\n",
    "\n",
    "library(dplyr)\n",
    "\n",
    "library(tidyverse)\n",
    "\n",
    "library(GEOquery)\n",
    "\n",
    "#read in data\n",
    "\n",
    "dat←read.csv(file=”../data/xxx.csv”)\n",
    "\n",
    "#dimension of database\n",
    "\n",
    "dim(dat)\n",
    "\n",
    "Rstudio"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[refine.bio](https://www.notion.so/refine-bio-ce044e3f338d499a944978c4b758a2b9?pvs=21)\n",
    "\n",
    "Alicia Mastretta en GitHub y en youtube, tiene varios videos de bioinformatica en castellano que me ayudaron al principio.\n",
    "Si quieres aprender bash/R/python tambien esta bien ver software carpentry.\n",
    "\n",
    "https://github.com/natapol/kitsune/blob/781eee227d6660f9e4808cf9a3b0c9029f743281/README.rst\n",
    "\n",
    "https://www.hyphy.org/installation/\n",
    "\n",
    "https://github.com/Stephane-S/Simplot_PlusPlus#Linux-UNIX-and-Mac-OS-versions\n",
    "\n",
    "hyphy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "[BeginnersGuide/Download - Python Wiki](https://wiki.python.org/moin/BeginnersGuide/Download)\n",
    "\n",
    "Installed Python 3.10.4 (Last stable version)\n",
    "\n",
    "brew install python\n",
    "\n",
    "add python to zsh: `echo \"alias python=/usr/bin/python3\" >> ~/.zshrc`\n",
    "\n",
    "his will configure your zsh profile to run `/usr/bin/python3`\n",
    " when `python`is run. If you are still facing issues, ensure that `python=$`\n",
    " where the $ sign should equal the path `python`is installed on.\n",
    "\n",
    "I obtained this: 11-16-22\n",
    "\n",
    "Python has been installed as\n",
    "\n",
    "/opt/homebrew/bin/python3\n",
    "\n",
    "Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to\n",
    "\n",
    "`python3`, `python3-config`, `pip3` etc., respectively, have been installed into\n",
    "\n",
    "/opt/homebrew/opt/python@3.10/libexec/bin\n",
    "\n",
    "You can install Python packages with\n",
    "\n",
    "pip3 install <package>\n",
    "\n",
    "They will install into the site-package directory\n",
    "\n",
    "/opt/homebrew/lib/python3.10/site-packages\n",
    "\n",
    "tkinter is no longer included with this formula, but it is available separately:\n",
    "\n",
    "brew install python-tk@3.10\n",
    "\n",
    "`echo \"alias python=/`opt/homebrew/bin/python3`\" >> ~/.zshrc`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://youtube.com/playlist?list=PLhR2Go-lh6X4fCAa3c_TBAjZD5A25s7fo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "https://github.com/hyginn/R-Intro\n",
    "\n",
    "[Introduction to R Session 1 - YouTube](https://www.youtube.com/watch?v=Mo5H_T0x9OE&list=PL3izGL6oi0S-h3swgRWK7qSxMHMamHemY)\n",
    "\n",
    "Notes:\n",
    "\n",
    "Start a new project\n",
    "\n",
    "init()\n",
    "\n",
    "Organizing files means giving them good names and placing them into folders where they are easy to find.\n",
    "\n",
    "When working with **R**, you need to specify its full name including the extension\n",
    "\n",
    "A path /Mac or Unix and \\Windows\n",
    "\n",
    "Top level directory in windows: :\\ \n",
    "\n",
    "Mac: / \n",
    "\n",
    "Avoid many subdirectories\n",
    "\n",
    "Try to save files as YYYY-MM-DD + version + content\n",
    "\n",
    "99- prefix keeps it sorted at the bottom of the directory listing\n",
    "\n",
    "Do not use spaces, hyphens or special characters in filenames.\n",
    "\n",
    "commandline interface\n",
    "\n",
    "MAC: Terminal\n",
    "\n",
    "Windows: cmd.exe (Cygwin creates a unix shell) https://en.wikipedia.org/wiki/Cygwin\n",
    "\n",
    "Package manager MAC: https://en.wikipedia.org/wiki/Homebrew\n",
    "\n",
    "https://github.com/Homebrew/brew/releases/tag/3.5.10\n",
    "\n",
    "`/bin/bash -c \"$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)\"`\n",
    "\n",
    "[https://brew.sh](https://brew.sh/)\n",
    "\n",
    "Windows: Chocolatey, scoop\n",
    "\n",
    "**A version control system:** https://github.com/gitster/git/blob/master/INSTALLhttps://git-scm.com/download/mac\n",
    "\n",
    "**LaTeX**\n",
    "\n",
    "https://www.benjack.io/2016/01/02/el-capitan-biocomputing.html\n",
    "\n",
    "I use the angle brackets only to indicate that you should not type **Year**\n",
    " literally, but substitute\n",
    " the correct value\n",
    "\n",
    "```\n",
    "Elcid Barret\n",
    "\n",
    "```\n",
    "\n",
    "... and not   your name   or   <Elcid Barret>\n",
    "\n",
    "The `>` character at the beginning of the line is always just **R'**s *input prompt*, it tells you that you can type something now - you never actually type `>` at the beginning of a line. If you read:\n",
    "\n",
    "`> getwd()`\n",
    "\n",
    "... you need to type:\n",
    "\n",
    "`getwd()`\n",
    "\n",
    "`[1]`\n",
    " or similar, this is **R'**\n",
    "s **output**\n",
    " on the console.[[3]](http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Installation#cite_note-3)\n",
    " The `#`\n",
    " character marks the following text as a comment which is not executed by **R**\n",
    ". These are lines that you **do not type**\n",
    ".\n",
    "\n",
    "- `/`  ◁ this is a **forward-slash**. It leans forward in the reading direction.\n",
    "- `\\`  ◁ this is a **backslash**. It leans backward in the reading direction.\n",
    "- `( )`  ◁ these are **parentheses**.\n",
    "- [ ]  `[ ]`  ◁ these are (square) **brackets**.\n",
    "- `< >`  ◁ these are angle brackets.\n",
    "- `{ }`  ◁ these are (curly) **braces**.\n",
    "- `\"`  ◁ this, and **only** this is a quotation mark or double quote. All of these are not: . They will break your code. Especially the first two are often automatically inserted by MSWord and hard to distinguish.\n",
    "    \n",
    "    “”„«»\n",
    "    \n",
    "    [[4]](http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Installation#cite_note-4)\n",
    "    \n",
    "- `'`  ◁ this, and **only** this is a single quote. All of these are not: . They will break your code. Especially the first two are often automatically inserted by MSWord and hard to distinguish.\n",
    "    \n",
    "    ‘’‚‹›\n",
    "    \n",
    "\n",
    "MSWord is not useful as a code editor.\n",
    "\n",
    "`getwd()`\n",
    "current working directory\n",
    "\n",
    "install.packages(\"<package-name\")\n",
    "\n",
    "- `packagename::function()` is usually the preferred idiom to use functionality that is providedby a package.\n",
    "- For some packages, or when you use a particular package **a lot**, you can \"load\" the package with `library(packagename)` (note: **no** quotation marks in this case.) Then you can use the functions simply by typing `function()`.\n",
    "\n",
    "install.packages(”ggplot2”)\n",
    "\n",
    "install.packages(”dplyr”)\n",
    "\n",
    "library(ggplot2)\n",
    "\n",
    "library(dplyr)\n",
    "\n",
    "[R Kadai](https://www.notion.so/R-Kadai-3036b756085148beaa34374db8a45bbc?pvs=21)\n",
    "\n",
    "[R 06-2023](https://www.notion.so/R-06-2023-49d7ff1dac5e44c2ba4c240a1ad9e6fa?pvs=21)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RIMA\n",
    "\n",
    "https://liulab-dfci.github.io/RIMA/how-to-run-rima.html\n",
    "\n",
    "https://hbctraining.github.io/Intro-to-shell-flipped/lessons/01_the_filesystem.html\n",
    "\n",
    "`$ls`\n",
    "\n",
    "List\n",
    "\n",
    "`$ls -F` detailed list\n",
    "\n",
    "`$ls -l` long list\n",
    "\n",
    "`$ls -lF` combination of l and f\n",
    "\n",
    "`$cp` `-r /n/groups/hbctraining/unix_lesson/ .`\n",
    "\n",
    "copy\n",
    "\n",
    "n/groups/hbctraining/unix_lesson/) and the location of the destination (.) #here\n",
    "\n",
    "`$cd`\n",
    "\n",
    "change of directory\n",
    "\n",
    "`$man ls` manual\n",
    "\n",
    "- ‘spacebar’ to go forward\n",
    "- ‘b’ to go backward\n",
    "- Up or down arrows to go forward or backward, respectively\n",
    "- q -> exit\n",
    "\n",
    "![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/2d5c04e8-0bca-41d3-aad5-cd1738597423/Untitled.png)\n",
    "\n",
    "`$pwd`**p**resent **w**orking **d**irectory\n",
    "\n",
    "*`~`full path of the directory*\n",
    "\n",
    "`$cd *~/*`\n",
    "\n",
    "`$cd ..`\n",
    "\n",
    "`$tree`  tree directory\n",
    "\n",
    "Use tab completion\n",
    "\n",
    "**A full path always starts with a `/`, a relative path does not.**\n",
    "\n",
    "cp path/to/item-being-copied path/to/new-copied-item\n",
    "\n",
    "`$ cp Mov10_oe_1.subset.fq Mov10_oe_1.subset-copy.fq`\n",
    "\n",
    "-r to copy structures and subdirectories\n",
    "\n",
    "`$ mkdir` name of the directory\n",
    "\n",
    "mv path/to/item-being-moved path/to/destination\n",
    "\n",
    "move or rename it\n",
    "\n",
    "```\n",
    "cd          # change directory\n",
    "ls          # list contents\n",
    "man         # manual for a command\n",
    "pwd         # check present working directory\n",
    "tree        # prints a tree of the file structure\n",
    "cp          # copy\n",
    "mkdir       # make new directory\n",
    "mv          # move or rename\n",
    "rm          # remove/delete\n",
    "```\n",
    "\n",
    "```\n",
    "~           # home directory\n",
    ".           # current directory\n",
    "..          # parent directory\n",
    "```\n",
    "\n",
    "https://hbctraining.github.io/Intro-to-shell-flipped/lessons/02_wildcards_shortcuts.html\n",
    "\n",
    "*everything\n",
    "\n",
    "`$ ls *****fq`\n",
    "\n",
    "This lists every file that ends with a `fq`. Try this command:\n",
    "\n",
    "`$ ls /usr/bin/*****.sh`\n",
    "\n",
    "This lists every file in `/usr/bin` directory that ends in the characters `.sh`. “*” can be placed anywhere in your pattern. For example:\n",
    "\n",
    "`$ ls Mov10*****fq`\n",
    "\n",
    "`ls /bin/d?`\n",
    "\n",
    "ls /bin/d??\n",
    "\n",
    "https://www.putorius.net/standard-wildcards-globbing-patterns-in.html\n",
    "\n",
    "**Home directory or “~”**\n",
    "\n",
    "`s ~`\n",
    ", `ls ~/.`\n",
    ", and `ls /home/username`\n",
    "\n",
    "You can also review your recent commands with the `history` command. Just enter:\n",
    "\n",
    "`$ history`\n",
    "\n",
    "Up and down\n",
    "\n",
    "### Cancel a command\n",
    "\n",
    "Sometimes as you enter a command, you realize that you don’t want to continue or run the current line. Instead of deleting everything you have entered (which could be very long), you could quickly cancel the current line and start a fresh prompt with Ctrl + C.\n",
    "\n",
    "**Other handy command-related shortcuts**\n",
    "\n",
    "- Ctrl +  will bring you to the start of the command you are writing.\n",
    "    \n",
    "    A\n",
    "    \n",
    "- Ctrl +  will bring you to the end of the command.\n",
    "    \n",
    "    E\n",
    "    \n",
    "    ```\n",
    "    ~           # home dir\n",
    "    .           # current dir\n",
    "    ..          # parent dir\n",
    "    *           # wildcard\n",
    "    ctrl + c    # cancel current command\n",
    "    ctrl + a    # start of line\n",
    "    ctrl + e    # end of line\n",
    "    history\n",
    "    ```\n",
    "    \n",
    "    ### `cat` command\n",
    "    \n",
    "    The easiest way to examine a file is to just print out all of its contents using the command `cat`. We can test this out by printing the contents of `~/unix_lesson/other/sequences.fa`. Enter the command followed by the filename, including the path when necessary:\n",
    "    \n",
    "    `$ cat ~/unix_lesson/other/sequences.fa`\n",
    "    \n",
    "    The `cat` command prints out the all the contents of `sequences.fa` to the screen.\n",
    "    \n",
    "    > cat stands for catenate; it has many uses and printing the contents of a files onto the terminal is one of them.\n",
    "    > \n",
    "    \n",
    "    `$ ls -lh ~/unix_lesson/raw_fastq`\n",
    "    \n",
    "    > The ls command has a modifier -h when paired with -l, will list the files and also print sizes of files in human readable format.\n",
    "    > \n",
    "    \n",
    "    `$ less Mov10_oe_1.subset.fq`\n",
    "    \n",
    "    Rather than printing to screen, the `less` command opens the file in a new buffer allowing you to navigate through it. Does this look familiar? You might remember encountering a similar interface when you used the `man` command. This is because `man` is using the `less` command to open up the documentation files! The keys used to move around the file are identical to the `man` command. Below we have listed some additional shortcut keys for naviagting through your file when using `less`.\n",
    "    \n",
    "    Shortcuts for `less`\n",
    "    \n",
    "    | key | action |\n",
    "    | --- | --- |\n",
    "    | SPACE | to go forward |\n",
    "    | b | to go backwards |\n",
    "    | g | to go to the beginning |\n",
    "    | G | to go to the end |\n",
    "    | q | to quit |\n",
    "    \n",
    "    Use the shortcut keys to move through your FASTQ file, we will explore these files in more detail later in the workshop.\n",
    "    \n",
    "    ### `ead` and `tail` commands\n",
    "    \n",
    "    There’s another way that we can look at files, and just look at part of them. In particular, if we just want to see the beginning or end of the file to see how it’s formatted.\n",
    "    \n",
    "    The commands are `head` and `tail` and they just let you look at the beginning and end of a file respectively.\n",
    "    \n",
    "    `$ head Mov10_oe_1.subset.fq`\n",
    "    \n",
    "    `$ tail Mov10_oe_1.subset.fq`\n",
    "    \n",
    "    By default, the first or last 10 lines will be printed to screen. The `-n` option can be used with either of these commands to specify the number `n` lines of a file to display. For example, let’s print the first/last line of the file:\n",
    "    \n",
    "    `$ head -n 1 Mov10_oe_1.subset.fq\n",
    "    \n",
    "    $ tail -n 1 Mov10_oe_1.subset.fq`\n",
    "    \n",
    "    G**raphical user interface (GUI) text editors**\n",
    "     and **command-line editors**\n",
    "    [BBEdit](http://www.barebones.com/products/bbedit/)\n",
    "    , [Sublime](http://www.sublimetext.com/)\n",
    "    , and [Notepad++](http://notepad-plus-plus.org/)\n",
    "    \n",
    "    ![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/9592188c-c43c-4501-953d-297fec971a78/Untitled.png)\n",
    "    \n",
    "    https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course/blob/master/resources/VI_CommandReference.pdf\n",
    "    \n",
    "    Vim has ***two basic modes*** that will allow you to create documents and edit your text:\n",
    "    \n",
    "    - ***command mode (default mode):*** will allow you to save and quit the program (and execute other more advanced commands).\n",
    "    - ***insert (or edit) mode:*** will allow you to write and edit text\n",
    "    \n",
    "    Let’s *change to insert mode*\n",
    "     by typing i\n",
    "    . **Note the `--INSERT--` at the bottom left hand of the screen.**\n",
    "    \n",
    "    After you have finished typing, **press esc to enter command mode.**\n",
    "    \n",
    "    **Note the `--INSERT--` has now disappeared from the bottom of the screen.**\n",
    "    \n",
    "    > \n",
    "    > \n",
    "    \n",
    "    ### Review of Vim modes\n",
    "    \n",
    "    | key | action |\n",
    "    | --- | --- |\n",
    "    | i | insert mode - to write and edit text |\n",
    "    | esc | command mode - to issue commands / shortcuts |\n",
    "    \n",
    "    ### Saving and Quitting\n",
    "    \n",
    "    To **“write to file”** or save the modifications made to the file, **type :w** when in command mode. You can see the commands you type in the bottom left-hand corner of the screen.\n",
    "    \n",
    "    Alternatively, we can **write to file (save changes) and quit**\n",
    "     all at once by **typing :wq**\n",
    "    . Now, you should have exited vim and returned back to your command prompt.\n",
    "    \n",
    "    q**uit without saving**\n",
    "     by going into command mode by pressing the esc\n",
    "     key, and then **typing :q!**\n",
    "    .\n",
    "    \n",
    "    | key (in command mode) | action |\n",
    "    | --- | --- |\n",
    "    | :w | to write to file (save) |\n",
    "    | :wq | to write to file and quit |\n",
    "    | :q! | to quit without saving |\n",
    "    \n",
    "    | key (in command mode) | action |\n",
    "    | --- | --- |\n",
    "    | :set number | to number lines |\n",
    "    | :set nonumber | to remove line numbers |\n",
    "    \n",
    "    | key (in command mode) | action |\n",
    "    | --- | --- |\n",
    "    | gg | to move to top of file |\n",
    "    | G | to move to bottom of file |\n",
    "    | $ | to move to end of line |\n",
    "    | 0 | to move to beginning of line |\n",
    "    | w | to move to next word |\n",
    "    | b | to move to previous word |\n",
    "    \n",
    "    Practice some of the editing shortcuts, then **quit the document without saving any changes**.\n",
    "    \n",
    "    **Editing the file**\n",
    "    \n",
    "    | key (in command mode) | action |\n",
    "    | --- | --- |\n",
    "    | dw | to delete word |\n",
    "    | dd | to delete line |\n",
    "    | u | to undo |\n",
    "    | Ctrl + r | to redo |\n",
    "    | /pattern | to search for a pattern (n/N to move to next/previous match) |\n",
    "    | :%s/search/replace/g | to search for a pattern and replace for all occurences |\n",
    "    \n",
    "    ## Searching files with `grep` command\n",
    "    \n",
    "    We went over how to search within a file using `less`. We can also search within files without even opening them, using `grep`. Simply put `grep` is a command-line utility for searching plain-text data sets for lines matching a pattern or **reg**ular **ex**pression (regex).\n",
    "    \n",
    "    **g**lobally search for a **r**egular **e**xpression and **p**rint matching lines (g/re/p).\n",
    "    \n",
    "    grep search_term filename\n",
    "    \n",
    "    F**ASTQ file format**\n",
    "    \n",
    "    | Line | Description |\n",
    "    | --- | --- |\n",
    "    | 1 | Read name preceded by ‘@’ |\n",
    "    | 2 | The actual DNA sequence |\n",
    "    | 3 | Read name (same as line 1) preceded by a ‘+’ or just a ‘+’ sign |\n",
    "    | 4 | String of characters which represent the quality score of each nucleotide in line 2; must have same number of characters as line 2 |\n",
    "    \n",
    "    looks like the `-B`\n",
    "     and `-A`\n",
    "     arguments for grep will be useful to return the matched line plus one before (`-B 1`\n",
    "    ) and two lines after (`-A 2`\n",
    "    ).\n",
    "    \n",
    "    [grep](https://www.notion.so/grep-9814b8c0795d4f528a5bbb49daea8042?pvs=21)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "# grep\n",
    "\n",
    "## Learning objectives\n",
    "\n",
    "- Search for characters or patterns in a text file using the `grep` command\n",
    "- Write to and append a file using output redirection\n",
    "- Use the pipe (`|`) character to chain together commands\n",
    "\n",
    "## Searching files with `grep` command\n",
    "\n",
    "We went over how to search within a file using `less`. We can also search within files without even opening them, using `grep`. Simply put `grep` is a command-line utility for searching plain-text data sets for lines matching a pattern or **reg**ular **ex**pression (regex).\n",
    "\n",
    "> Why the word “grep”? It is a shortened form of globally search for a regular expression and print matching lines (g/re/p).\n",
    "> \n",
    "\n",
    "The syntax for `grep` is as follows: `grep search_term filename`. The pattern that we want to search is specified in `search_term` slot, and the file we want to search within is specified in the `filename` slot. Let’s give it a try by searching the FASTQ files in the `raw_fastq` directory.\n",
    "\n",
    "FASTQ files contain the sequencing reads (nucleotide sequences) output from a sequencing facility. Each sequencing read in a FASTQ file is associated with four lines, with the first line (header line) always starting with an `@` symbol. A whole FASTQ record for a single read should appear similar to the following:\n",
    "\n",
    "`@HWI-ST330:304:H045HADXX:1:1101:1111:61397\n",
    "CACTTGTAAGGGCAGGCCCCCTTCACCCTCCCGCTCCTGGGGGANNNNNNNNNNANNNCGAGGCCCTGGGGTAGAGGGNNNNNNNNNNNNNNGATCTTGG\n",
    "+\n",
    "B?@DDDDDDHHH?GH:?FCBGGB@C?DBEGIIIIAEF;FCGGI#########################################################`\n",
    "\n",
    "> More information about the FASTQ file format\n",
    "> \n",
    "> \n",
    "> \n",
    "> | Line | Description |\n",
    "> | --- | --- |\n",
    "> | 1 | Read name preceded by ‘@’ |\n",
    "> | 2 | The actual DNA sequence |\n",
    "> | 3 | Read name (same as line 1) preceded by a ‘+’ or just a ‘+’ sign |\n",
    "> | 4 | String of characters which represent the quality score of each nucleotide in line 2; must have same number of characters as line 2 |\n",
    "> \n",
    "> You can find more information about FASTQ files in [this lesson](https://hbctraining.github.io/Intro-to-rnaseq-hpc-salmon-flipped/lessons/05_qc_running_fastqc_interactively.html) from our [RNA-seq workshop](https://hbctraining.github.io/Intro-to-rnaseq-hpc-salmon-flipped/).\n",
    "> \n",
    "\n",
    "Suppose we want to see how many reads in our file `Mov10_oe_1.subset.fq` contain “bad” data, i.e. reads with 10 consecutive Ns (`NNNNNNNNNN`).\n",
    "\n",
    "`$ cd ~/unix_lesson/raw_fastq\n",
    "\n",
    "$ grep NNNNNNNNNN Mov10_oe_1.subset.fq`\n",
    "\n",
    "We get back a lot of reads or lines of text!\n",
    "\n",
    "What if we wanted to see the whole FASTQ record for each of these reads? We would need to modify the default behavior of `grep` and specify some argument/options. To look for all available options for the `grep` command, we can type `grep --help` (or `man grep`).\n",
    "\n",
    "Looks like the `-B` and `-A` arguments for grep will be useful to return the matched line plus one before (`-B 1`) and two lines after (`-A 2`). Since each record is four lines, using these arguments will return the whole record. Within the whole record, the second line will be the actual sequence that has the pattern we searched for.\n",
    "\n",
    "`$ grep -B 1 -A 2 NNNNNNNNNN Mov10_oe_1.subset.fq`\n",
    "\n",
    "`@HWI-ST330:304:H045HADXX:1:1101:1111:61397\n",
    "CACTTGTAAGGGCAGGCCCCCTTCACCCTCCCGCTCCTGGGGGANNNNNNNNNNANNNCGAGGCCCTGGGGTAGAGGGNNNNNNNNNNNNNNGATCTTGG\n",
    "+\n",
    "@?@DDDDDDHHH?GH:?FCBGGB@C?DBEGIIIIAEF;FCGGI#########################################################\n",
    "--\n",
    "@HWI-ST330:304:H045HADXX:1:1101:1106:89824\n",
    "CACAAATCGGCTCAGGAGGCTTGTAGAAAAGCTCAGCTTGACANNNNNNNNNNNNNNNNNGNGNACGAAACNNNNGNNNNNNNNNNNNNNNNNNNGTTGG\n",
    "+\n",
    "?@@DDDDDB1@?:E?;3A:1?9?E9?<?DGCDGBBDBF@;8DF#########################################################`\n",
    "\n",
    "---\n",
    "\n",
    "**Exercises**\n",
    "\n",
    "1. Search for the sequence CTCAATGAGCCA in `Mov10_oe_1.subset.fq`. How many sequences do you find?\n",
    "2. In addition to finding the sequence, how can you modify the command so that your search also returns the name of the sequence?\n",
    "3. If you want to search for that sequence in **all** Mov10 replicate fastq files, what command would you use?\n",
    "    - ***Answers***\n",
    "\n",
    "---\n",
    "\n",
    "### More about searching text files with `grep`\n",
    "\n",
    "### Group separators (`-`), and how to remove them\n",
    "\n",
    "You will notice that when we use the `-B` and/or `-A` arguments with the `grep` command, the output has some additional lines with dashes (`--`), these dashes work to separate your returned “groups” of lines and are referred to as “group separators”. This might be problematic if you are trying to maintain the FASTQ file structure or if you simply do not want them in your output. Using the argument `--no-group-separator` with `grep` will disable this behavior:\n",
    "\n",
    "`$ grep -B 1 -A 2 --no-group-separator NNNNNNNNNN Mov10_oe_1.subset.fq`\n",
    "\n",
    "Now your output should be returned as:\n",
    "\n",
    "`@HWI-ST330:304:H045HADXX:1:1101:1111:61397\n",
    "CACTTGTAAGGGCAGGCCCCCTTCACCCTCCCGCTCCTGGGGGANNNNNNNNNNANNNCGAGGCCCTGGGGTAGAGGGNNNNNNNNNNNNNNGATCTTGG\n",
    "+\n",
    "@?@DDDDDDHHH?GH:?FCBGGB@C?DBEGIIIIAEF;FCGGI#########################################################\n",
    "@HWI-ST330:304:H045HADXX:1:1101:1106:89824\n",
    "CACAAATCGGCTCAGGAGGCTTGTAGAAAAGCTCAGCTTGACANNNNNNNNNNNNNNNNNGNGNACGAAACNNNNGNNNNNNNNNNNNNNNNNNNGTTGG\n",
    "+\n",
    "?@@DDDDDB1@?:E?;3A:1?9?E9?<?DGCDGBBDBF@;8DF#########################################################`\n",
    "\n",
    "### Which line number has a match?\n",
    "\n",
    "Another useful option when using `grep` is the `-n` option, which will print out the line number from the file for the match. Adding this option to our previous command would work like this:\n",
    "\n",
    "`$ grep -B 1 -A 2 --no-group-separator -n NNNNNNNNNN Mov10_oe_1.subset.fq`\n",
    "\n",
    "This would return the output:\n",
    "\n",
    "`861213-@HWI-ST330:304:H045HADXX:1:1101:1111:61397\n",
    "861214:CACTTGTAAGGGCAGGCCCCCTTCACCCTCCCGCTCCTGGGGGANNNNNNNNNNANNNCGAGGCCCTGGGGTAGAGGGNNNNNNNNNNNNNNGATCTTGG\n",
    "861215-+\n",
    "861216-@?@DDDDDDHHH?GH:?FCBGGB@C?DBEGIIIIAEF;FCGGI#########################################################\n",
    "861953-@HWI-ST330:304:H045HADXX:1:1101:1106:89824\n",
    "861954:CACAAATCGGCTCAGGAGGCTTGTAGAAAAGCTCAGCTTGACANNNNNNNNNNNNNNNNNGNGNACGAAACNNNNGNNNNNNNNNNNNNNNNNNNGTTGG\n",
    "861955-+\n",
    "861956-?@@DDDDDB1@?:E?;3A:1?9?E9?<?DGCDGBBDBF@;8DF#########################################################`\n",
    "\n",
    "A small thing you should note is that when using the `-n` option, lines that have a `:` after the line number correspond to the lines with the match (e.g `861214:CACTTGTAAGGGCAGGCCCCCTTCACCCTCCCGCTCCTGGGGGANNNNNNNNNNANNN...`), while lines with a `-` after the line number are the surrounding lines retrieved when using the `-A` and/or `-B` options (e.g. `861213-@HWI-ST330:304:H045HADXX:1:1101:1111:61397`).\n",
    "\n",
    "### Only returning lines that **DO NOT** match\n",
    "\n",
    "One last `grep` option you might find quite useful is the `-v` option, which does an inverted match. This will return everything that does ***not*** match the pattern. In order to demonstrate this let’s first view a smaller file that you have.\n",
    "\n",
    "`$ cat ~/unix_lesson/other/Mov10_rnaseq_metadata.txt`\n",
    "\n",
    "This is the metadata file for the FASTQ data we are looking at. This should return:\n",
    "\n",
    "`sample\tcelltype\n",
    "OE.1\tMov10_oe\n",
    "OE.2\tMov10_oe\n",
    "OE.3\tMov10_oe\n",
    "IR.1\tnormal\n",
    "IR.2\tnormal\n",
    "IR.3\tnormal`\n",
    "\n",
    "Now, let’s consider the case that we didn’t want to output the “normal” cell type. We can use the `-v` option in `grep` like this:\n",
    "\n",
    "`$ grep -v normal ~/unix_lesson/other/Mov10_rnaseq_metadata.txt`\n",
    "\n",
    "This will return all of lines that don’t have “normal” in the line.\n",
    "\n",
    "`sample\tcelltype\n",
    "OE.1\tMov10_oe\n",
    "OE.2\tMov10_oe\n",
    "OE.3\tMov10_oe`\n",
    "\n",
    "## Redirection\n",
    "\n",
    "When we use `grep`, the matching lines print to the Terminal (also called Standard Output or “stdout”). If the result of the `grep` search is a few lines, we can view them easily, but if the output is very long, the lines will just keep printing and we won’t be able to see anything except the last few lines. You have experienced this when you searched for the pattern `NNNNNNNNNN`. How can we capture them instead?\n",
    "\n",
    "We can do that with something called “redirection”. The idea is that we’re redirecting the output from the Terminal (all the stuff that went whizzing by) to something else. In this case, we want to save it to a file, so that we can look at it later.\n",
    "\n",
    "### Redirecting with “>”\n",
    "\n",
    "**The redirection command for writing something to file is `>`.**\n",
    "\n",
    "Let’s try it out and put all the sequences that contain ‘NNNNNNNNNN’ from the `Mov10_oe_1.subset.fq` into another file called `bad_reads.txt`.\n",
    "\n",
    "`$ grep -B 1 -A 2 NNNNNNNNNN Mov10_oe_1.subset.fq **>** bad_reads.txt`\n",
    "\n",
    "The prompt will go away for a little bit and then you will get it back, but nothing will be printed on the Terminal. But, you should have a new file called `bad_reads.txt`!\n",
    "\n",
    "`$ ls -l`\n",
    "\n",
    "Take a look at the file and see if it contains what you think it should.\n",
    "\n",
    "> NOTE: If we already had a file named bad_reads.txt in our directory, it would have overwritten it without any warning!\n",
    "> \n",
    "\n",
    "### Redirecting (and appending) with “»”\n",
    "\n",
    "**The redirection command for appending something to an existing file is `>>`.**\n",
    "\n",
    "If we use `>>`, it will append to the existing content in a file, rather than overwrite it. This can be useful for saving more than one search. For example, the following command will append the bad reads from Mov10_oe_2 to the bad_reads.txt file that we just generated.\n",
    "\n",
    "`$ grep -B 1 -A 2 NNNNNNNNNN Mov10_oe_2.subset.fq **>>** bad_reads.txt\n",
    "\n",
    "$ ls -l`\n",
    "\n",
    "Did the size of the `bad_reads.txt` file change?\n",
    "\n",
    "Since our `bad_reads.txt` file isn’t a raw_fastq file, we should move it to a different location within our directory. Let’s move it to the `other` folder using the command `mv`.\n",
    "\n",
    "`$ mv bad_reads.txt ../other/`\n",
    "\n",
    "### Redirecting with pipes “|” (or piping)\n",
    "\n",
    "**The redirection command for using the output of a command as input for a different command is `|`.**\n",
    "\n",
    "**The pipe key** (|) is very likely not something you use very often (it is on the same key as the back slash (\\\\), right above the Enter/Return key).\n",
    "\n",
    "What `|` does is take the output from one command, e.g. the output from `grep` that went whizzing by and runs it through the command specified after it. When it was all whizzing by before, we wished we could just take a look at it! Maybe we could use `less` instead of the rapid scroll. Well, it turns out that we can! We can **pipe the output `grep` command** to `less` to slowly scroll through, or to `head` to just see the first few lines.\n",
    "\n",
    "`$ grep -B 1 -A 2 NNNNNNNNNN Mov10_oe_1.subset.fq | less`\n",
    "\n",
    "Now we can use the arrows to scroll up and down and use `q` to get out.\n",
    "\n",
    "Or we could just take a glance to see what the output looks like.\n",
    "\n",
    "`$ grep -B 1 -A 2 NNNNNNNNNN Mov10_oe_1.subset.fq | head -n 5`\n",
    "\n",
    "Another thing we can also do is count the number of lines output by `grep`.\n",
    "\n",
    "The `wc` command stands for ***w**ord **c**ount*. This command counts the number of lines, words and characters in the text input given to it. The `-l` argument will only count the number of lines instead of counting everything.\n",
    "\n",
    "`$ grep NNNNNNNNNN Mov10_oe_1.subset.fq | wc -l`\n",
    "\n",
    "*Try it out without the `-l` to see the full output.*\n",
    "\n",
    "> Tip - Similar to grep, you can type wc --help or man wc to see all options.\n",
    "> \n",
    "\n",
    "**About Pipes:**\n",
    "\n",
    "- The pipe is a very important/powerful concept in Shell\n",
    "- You can string along as many commands together as you like\n",
    "\n",
    "The philosophy behind the three redirection operators (`>`, `>>`, `|`) you have learned so far is that none of them by themselves do a lot. BUT when you start chaining them together, you can do some really powerful things really efficiently.\n",
    "\n",
    "**To be able to use the shell effectively, becoming proficient in the use of the pipe and redirection operators is essential.**\n",
    "\n",
    "[piping](https://www.notion.so/piping-78cdb51922fb4eb8876399855489a89f?pvs=21)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://trumpexcel.com/insert-blank-row-after-every-row/\n",
    "\n",
    "bin/iqtree2 -s NCBI_seq_HTLV_Aligned_22-5-22.aln -m TIM2+I+G -B 1000 -nt 7 --undo\n",
    "\n",
    "bin/iqtree2 -s NCBI_seq_HTLV_Aligned_22-5-22 -m MF -mtree\n",
    "\n",
    "bin/iqtree2 -s Peru_sequences_cg_HTLV_22-5-22_orden_aligned.aln -m MF -mtree -nt 7 --undo\n",
    "\n",
    "bin/iqtree2 -s total_seq_aligned.aln -m MF -mtree -nt 7 --undo\n",
    "\n",
    "bin/iqtree2 -s total_seq_aligned.aln -m TIM2+I+G -B 1000 -nt 7 --undo\n",
    "\n",
    "https://twitter.com/fpozoca/status/1531926725369659392?s=21&t=r5UoSQm1q5G09WOaWW6UMQ\n",
    "\n",
    "https://phylo.bio.ku.edu/content/biol-848-phylogenetic-methods-10\n",
    "\n",
    "https://phylo.bio.ku.edu/content/biol-848-phylogenetic-methods-10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "https://uoa-eresearch.github.io/eresearch-cookbook/recipe/2014/11/20/conda/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "http://steipe.biochemistry.utoronto.ca/abc/index.php/R_tutorial\n",
    "\n",
    "https://urban-institute.medium.com/how-to-create-state-and-county-maps-easily-in-r-577d29300bb2https://github.com/rmcondor/rmaps-peruvian-case\n",
    "https://github.com/Cibiv/IQ-TREE\n",
    "https://github.com/musajajorge/mapsPERU"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[An Introduction to Phylogenomics](https://astrobiomike.github.io/genomics/phylogenomics)\n",
    "\n",
    "First question: Genome level versus gene level, multiple concatenated genes\n",
    "\n",
    "Single-copy core genes\n",
    "\n",
    "Multiple gene-copies (paralogs)\n",
    "\n",
    "1. input genomes\n",
    "2. align individual target genes\n",
    "3. stick alignments together\n",
    "4. infer evolutionary relationships\n",
    "\n",
    "https://merenlab.org/data/\n",
    "\n",
    "https://github.com/chrishah/phylogenomics-intro\n",
    "\n",
    "https://busco.ezlab.org/\n",
    "\n",
    "https://evomics.org/learning/phylogenetics/beast/viruses/\n",
    "\n",
    "https://jlsteenwyk.com/resources.html\n",
    "\n",
    "https://docs.nextstrain.org/en/latest/tutorials/creating-a-workflow.html\n",
    "\n",
    "[Home · JimmyNkaiwuatei/Phygenetic-analysis-of-West-Nile-Virus-WNV-outbreak-in-Andalussia-spain Wiki · GitHub](https://github.com/JimmyNkaiwuatei/Phygenetic-analysis-of-West-Nile-Virus-WNV-outbreak-in-Andalussia-spain/wiki)\n",
    "\n",
    "Geographic according to Peru\n",
    "\n",
    "Review again PCA \n",
    "\n",
    "Classify according Region (PCA)\n",
    "\n",
    "Qhich country they are close to?\n",
    "\n",
    "https://www.hiv.lanl.gov/content/sequence/HIV/HIVTools.html\n",
    "\n",
    "https://github.com/Brjacob/iGEM_UGent_2020\n",
    "\n",
    "Conda update anaconda\n",
    "\n",
    "1. Instalar Homebrew\n",
    "\n",
    "**Instalar las herramientas de línea de comandos de Xcode**\n",
    "\n",
    "`xcode-select --install`\n",
    "\n",
    "Instalar homebrew\n",
    "\n",
    "`/bin/bash -c \"$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)\"`\n",
    "\n",
    "Forma manual\n",
    "\n",
    "`curl -fsSL -o install.sh https://raw.githubusercontent.com/Homebrew/install/master/install.sh`\n",
    "\n",
    "`less install.sh`\n",
    "\n",
    "`/bin/bash install.sh`\n",
    "\n",
    "[Phylogenetics](https://www.notion.so/Phylogenetics-e84a7e84a5af42f5bf18758567024850?pvs=21)\n",
    "\n",
    "Niema Moshiri\n",
    "\n",
    "[Niema Moshiri](https://www.notion.so/Niema-Moshiri-5122fa0fee114bc79cba6699c9e6c323?pvs=21)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
