{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "https://ubco-biology.github.io/BIOL202/\n",
    "\n",
    "\n",
    "https://towardsdatascience.com/my-workflow-for-open-and-reproducible-science-as-an-academic-researcher-in-biomedicine-b41eaabcd420\n",
    "\n",
    "https://cyverse-snakemake-vice.readthedocs-hosted.com/en/latest/\n",
    "\n",
    "https://nbis-reproducible-research.readthedocs.io/en/course_2104/git/\n",
    "\n",
    "\n",
    "Webinar: Workflow Management Using Snakemake\n",
    "\n",
    "\n",
    "\n",
    "https://anopheles-genomic-surveillance.github.io/workshop-3/about.html\n",
    "Demonstration of R to analyze COVID 19 data using Google's CoLab\n",
    "\n",
    "\n",
    "\n",
    "Snakemake #1 Writing your first Snakefile\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "]\n",
    "\n",
    "\n",
    "Webinar: Workflow Management Using Snakemake\n",
    "\n",
    "\n",
    "\n",
    "Biological Data Analytics Specialization\n",
    "\n",
    "\n",
    "First find a sequence in blast\n",
    "Spades for genome assembly\n",
    "Prokka (comparing against existing annitations)\n",
    "\n",
    "global and local alignment\n",
    "mafft vs muscle vs clustal omega\n",
    "(too slow for long sequences)\n",
    "reference guided multiple sequence alignment\n",
    "\n",
    "fasttree\n",
    "\n",
    "quast work with spades\n",
    "\n",
    "read mappers:\n",
    "minimap2\n",
    "unimap\n",
    "BWA\n",
    "bowtie2\n",
    "etc\n",
    "\n",
    "to obtain the consensus sequence \n",
    "an amplicon-based sequencing framework for accurately measuring intrhost virus diversity using primalseq and ivar\n",
    "\n",
    "ViralMSA (minimap2 is a good option)\n",
    "\n",
    "Markov chain\n",
    "probability of the next state given the current state\n",
    "\n",
    "dating a phylogeny\n",
    "-treetime ancestral state reconstruction\n",
    "-LSD2\n",
    "-logdate\n",
    "\n",
    "lineage or strain\n",
    "-pangolin: decision tree\n",
    "nextstrain\n",
    "covariants: mutations and lineage over time\n",
    "outbreak.info deaths, prevalence, etc\n",
    "\n",
    "\n",
    "to find clusters\n",
    "HIV trace transmission cluster engine\n",
    "treecluster given a tree\n",
    "\n",
    "\n",
    "https://github.com/SoyeonA/aBayesQR\n",
    "\n",
    "\n",
    "https://github.com/mmatschiner\n",
    "\n",
    "\n",
    "# Phylogenetics\n",
    "\n",
    "https://github.com/amkozlov/raxml-ng\n",
    "\n",
    "raxml-ng --all --msa 0823peruref_mafftaligned.fasta --model GTR+G --prefix T2 --threads 5\n",
    "\n",
    "\n",
    "raxml-ng --all --msa testAA.fa --model LG+G8+F --tree pars{10} --bs-trees 200\n",
    "\n",
    "Optimize branch lengths and free model parameters on a fixed topology\n",
    "\n",
    "     (using multiple partitions with proportional branch lengths)\n",
    "\n",
    "     ./raxml-ng --evaluate --msa testAA.fa --model partitions.txt --tree test.tree --brlen scaled\n",
    "\n",
    "https://github.com/amkozlov/raxml-ng/wiki/Output:-files-and-settings\n",
    "\n",
    "http://biopython.org/DIST/docs/tutorial/Tutorial.html\n",
    "\n",
    "https://www.sciencedirect.com/science/article/abs/pii/S235230262300087X\n",
    "\n",
    "http://tree.bio.ed.ac.uk/\n",
    "https://www.hiv.lanl.gov/content/sequence/HIV/links.html\n",
    "\n",
    "\n",
    "https://www.bali-phy.org/README.xhtml#ancestral_sequence_reconstruction\n",
    "\n",
    "https://ben-redelings.org/\n",
    "\n",
    "https://www.bali-phy.org/Tutorial4.html\n",
    "https://www.bali-phy.org/\n",
    "\n",
    "\n",
    "git clone https://github.com/veg/hyphy hyphy\n",
    "cd hyphy\n",
    "cmake .\n",
    "make\n",
    "#para salir\n",
    "control +c\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "conda env list\n",
    "#to activate pip2\n",
    "conda activate pipdev2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "https://uc-r.github.io/hc_clustering#kmeans_clustering\n",
    "\n",
    "Matplotlib library\n",
    "https://bio-bwa.sourceforge.net/\n",
    "PacBio subreads were assembled into contigs using the PacBio de novo assembler Hierarchical Genome Assembly Process v4.0 (HGAP4)\n",
    "http://sanger-pathogens.github.io/iva/\n",
    "https://github.com/pauloluniyi/VGEA\n",
    "https://github.com/dnanexus-archive/viral-ngs/wiki\n",
    "https://viral-ngs.readthedocs.io/en/latest/reports.html\n",
    "https://github.com/rcs333/VAPiD\n",
    "https://academic.oup.com/bioinformatics/article/31/14/2374/254470\n",
    "https://www.sciencedirect.com/science/article/pii/S1567134819303065?pes=vor\n",
    "https://github.com/myl446/HivStudy\n",
    "https://www.mdpi.com/1099-4300/22/2/255"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
