{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8d7e1cd9-9a17-4092-8f36-226f77612dbc",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "https://omicstutorials.com/mastering-bioinformatics-analysis-with-fasta-sequences-a-biologists-guide-to-unix-and-linux/\n",
    "Bioconductor: https://www.bioconductor.org/help/course-materials/2015/EMBO/A01_RBiocForSequenceAnalysis.html \n",
    "basic commands: cd,ls,pwd,mkdir,rmdir,rm\n",
    "text editors:\n",
    "-nano\n",
    "-vim\n",
    "* also file permissions:\n",
    "  chmod (change mode to allow you to modify file permissions)\n",
    "  chmod u+x file\n",
    "  chmode 644 file\n",
    "  chown (change owner)\n",
    "  chown user:group file\n",
    "#### fasta format, begins with \">\"+ identifiers, then nucleotides or aminoacids\n",
    "* to visualize: cat (all the file), less or more (for large files), use q to exit\n",
    "* to combine fasta files: cat file1 file2 > filet\n",
    "* to split based on the identifiers: use AWK\n",
    " * `awk '/^>/{f=($0\".fasta\");} f{print>f}'input.fasta`\n",
    "* to extract sequences: AWK or GREP\n",
    " * `awk -v RS='>' -v seq=\"Sequencetarget\" '$1==seq{print \">\"$0}' input.fasta > output.fasta`\n",
    " * `grep -A 1 \"Sequencetarget\" input.fasta > output.fasta` #A -1 include one line after matching the sequences \n",
    "* to download data\n",
    " * `wget -o output.fata \"link\"`\n",
    " * `curl -o output.fasta \"link\"`\n",
    "* to verify quality control after download\n",
    " * `md5sum file.fastq`\n",
    " * `sha256sum file.fastq`\n",
    "* counting sequences, lengths, statistics (AWK)\n",
    " * `awk '/^>/{count++} END {print \"Number of Sequences:\" count}' file.fasta`\n",
    "* calculating average sequence length\n",
    " *  `awk '/^>/{total += length(sequence); count++} {sequence = $0} END{print \"average length: \" total/count}' file.fasta`\n",
    "\n",
    "#### Sequence Manipulation\n",
    "* Filtering\n",
    " * By sequence length:\n",
    "  * `awk '!?/^>/ {sequence=sequence $0+ next} length(sequence) >100 && length(sequence)<=200 {print $0; print sequence} {sequence=\"\"}' input.fasta > output.fasta`\n",
    " * By GC content:\n",
    "  * `awk '!/^>/ {sequence = sequence $0; next} {gc_count=gsub(/[ATGCatgc]/, \"\",sequence); } (gc_count/length(sequence)) >=0.6 && (gc_count / length(sequence)) <= 1.1 {print $0; print sequence} {sequence=\"\"}' input.fasta > output.fasta`\n",
    "\n",
    "\n",
    "* Format conversion\n",
    " * Seqtk\n",
    "  * `seqtk seq -a input.fastq > output.fa`\n",
    " * Bioawk\n",
    "#### Sequence alignment\n",
    "* tools like Bowtie2, BWA, STAR\n",
    "* indexing a reference genome\n",
    "* generate SAM/BAM files (bowtie)\n",
    "* to convert sam to bam (samtools)\n",
    "* to sort and infex a bam (samtools)\n",
    "\n",
    "#### Sequence analysis\n",
    "* BLAST\n",
    "* HMMER\n",
    "* ClustalW/MUSCLE\n",
    "* MEME: find transcription factors\n",
    "* GATK: variant calling SNP\n",
    "#### Analysis\n",
    "* Homology search\n",
    "* Motif finding\n",
    "* Variant calling\n",
    "* Phylogenetics\n",
    "* Functional Annotation\n",
    "* Comparative genomics\n",
    "* Structural bioinformatics\n",
    "#### Data visualization\n",
    "* R: ggplot2, pheatmap, plotly, bioconductor\n",
    "* Python: matopotlib,seaborn, plotly\n",
    "* Plotting sequences:\n",
    " * IGV\n",
    " * BAMview\n",
    " * JBrowse\n",
    "* Plotting coverage\n",
    " * BEDtools\n",
    " * Gviz (R)\n",
    " * Python libraries\n",
    "#### bash scripts\n",
    "`!/bin/bash\n",
    "#This is a simple bash script\n",
    "echo \"hello friends!\"`\n",
    "to make it executable `chmod +x myscript.sh`\n",
    "then run `./my_script.sh`\n",
    "#### Loops\n",
    "for all\n",
    "* `for file in *.txt;do\n",
    "echo \"processing file: $file\"\n",
    "done`\n",
    "* `find . -type f | while read -r file; do\n",
    "  echo \"processing file: $file\"\n",
    "  done`\n",
    "#### Continionals\n",
    "* `if [-e file.txt]; then\n",
    "echo \"File exist.\"\n",
    "else\n",
    "echo \"File does not exist.\"\n",
    "fi`\n",
    "*  `file_size=$(stat -c %s file.txt)\n",
    "  if [$file_size -gt 100]; then\n",
    "echo \"File is larger than 100 bytes.\"\n",
    "else\n",
    "echo \"File is smaller than or equal to 100 bytes\"\n",
    "fi`\n",
    "*`if [-d directory]; then\n",
    "echo \"It is a directory.\"\n",
    "else\n",
    "echo \"It's not a directory\"\n",
    "fi`\n",
    "#### Git\n",
    "* repository (repo), commit, branch, pull request.\n",
    "* `git init`\n",
    "* `git add`\n",
    "* `git commit -m \"initial commit\"`\n",
    "* `git remote add origin <url>`\n",
    "* `git branch -M main`\n",
    "* `git push -u origin main`\n",
    "#### Best practices\n",
    "* data backup, file naming data annotation, data directory structure, data sharing.\n",
    "* keep detailed records, use version control, reproducibility, data provenance (snakemake or neftflow)\n",
    "* sources\n",
    " * galaxy\n",
    " * bioconda\n",
    " * bioconductor\n",
    " * databases\n",
    " * forums\n",
    " *  \n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "be431c80-131d-4be7-a35c-e11c22a56fbf",
   "metadata": {},
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf7319eb-93aa-4b02-921f-a5ca9238996a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
